<!doctype html>



  


<html class="theme-next muse use-motion" lang="default">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>



  
  
  <link rel="stylesheet" media="all" href="/lib/Han/dist/han.min.css?v=3.3">




<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="b12Zu40L63PLgCtcn0IW8KCNRolmyQlaF22tf4rRLEg" />













  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="http://fonts.useso.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.1" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="BEDTools," />





  <link rel="alternate" href="/atom.xml" title="hope" type="application/atom+xml" />




  <link rel="shortcut icon" type="image/x-icon" href="/images/favicon.ico?v=5.1.1" />






<meta name="description" content="概述BEDTools是可用于genomic features的比较，相关操作及进行注释的工具。而genomic features通常使用Browser Extensible Data (BED) 或者 General Feature Format (GFF)文件表示，用UCSC Genome Browser进行可视化比较。
与BEDTools使用相关的基本概念已有的一些genome feature">
<meta property="og:type" content="article">
<meta property="og:title" content="bedtools 使用小结">
<meta property="og:url" content="http://tiramisutes.github.io/2016/03/18/bedtools.html">
<meta property="og:site_name" content="hope">
<meta property="og:description" content="概述BEDTools是可用于genomic features的比较，相关操作及进行注释的工具。而genomic features通常使用Browser Extensible Data (BED) 或者 General Feature Format (GFF)文件表示，用UCSC Genome Browser进行可视化比较。
与BEDTools使用相关的基本概念已有的一些genome feature">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/bed1.jpg">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/bed2.jpg">
<meta property="og:updated_time" content="2016-11-04T13:36:24.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="bedtools 使用小结">
<meta name="twitter:description" content="概述BEDTools是可用于genomic features的比较，相关操作及进行注释的工具。而genomic features通常使用Browser Extensible Data (BED) 或者 General Feature Format (GFF)文件表示，用UCSC Genome Browser进行可视化比较。
与BEDTools使用相关的基本概念已有的一些genome feature">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":0,"b2t":false,"scrollpercent":false},
    fancybox: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://tiramisutes.github.io/2016/03/18/bedtools.html"/>





  <title>bedtools 使用小结 | hope</title>
</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?08b17c69c7ba460d9ab5f188a048ee6d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>











  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">hope</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <h1 class="site-subtitle" itemprop="description">A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</h1>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      
        
        <li class="menu-item menu-item-code">
          <a href="/code" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-code"></i> <br />
            
            Code
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-footprint">
          <a href="/footprint" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-image"></i> <br />
            
            Footprint
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            Search
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal " itemscope itemtype="http://schema.org/Article">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/18/bedtools.html">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/author.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope">
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">bedtools 使用小结</h2>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2016-03-18T16:29:32+08:00">
                2016-03-18
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index">
                    <span itemprop="name">Bioinformatics</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
            <!--noindex-->
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2016/03/18/bedtools.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count hc-comment-count" data-xid="2016/03/18/bedtools.html" itemprop="commentsCount"></span>
                </a>
              </span>
              <!--/noindex-->
            
          

          
          
             <span id="/2016/03/18/bedtools.html" class="leancloud_visitors" data-flag-title="bedtools 使用小结">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Visitors </span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-wordcount">
              
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article</span>
                
                <span title="Words count in article">
                  4,114
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time</span>
                
                <span title="Reading time">
                  23
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        <h2 id="概述">概述</h2><p>BEDTools是可用于genomic features的比较，相关操作及进行注释的工具。而genomic features通常使用Browser Extensible Data (BED) 或者 General Feature Format (GFF)文件表示，用UCSC Genome Browser进行可视化比较。</p>
<h2 id="与BEDTools使用相关的基本概念">与BEDTools使用相关的基本概念</h2><p>已有的一些genome features信息一般由BED格式或者GFF格式进行存储。<br>genome features: 功能元素（gene）， 遗传多态性 (SNPs, INDELs, or structural variants), 已经由测序或者其他方法得到的注释信息，也可以是自定义的一些特征信息。<br>Overlapping/intersecting features: 两个genome features的区域至少有一个bp的共同片段。</p>
<h2 id="BED和GFF文件的一个差异">BED和GFF文件的一个差异</h2><p><strong>BED文件中起始坐标为0，结束坐标至少是1,； GFF中起始坐标是1而结束坐标至少是1。</strong></p>
<h2 id="相关格式">相关格式</h2><h3 id="BED_format">BED format</h3><p>BEDTools主要使用BED格式的前三列，BED可以最多有12列。BED格式的常用列描述如下：<br>chrom: 染色体信息， 如chr1, III, myCHrom, contig1112.23, 必须有<br>start: genome feature的起始位点，从0开始， 必须有<br>end: genome feature的终止位点，至少为1， 必须有<br>score: 可以是p值等等一些可以刻量化的数值信息<br>strands: 正反链信息</p>
<h3 id="GFF_format">GFF format</h3><p><strong>seqname</strong> - name of the chromosome or scaffold; chromosome names can be given with or without the ‘chr’ prefix. <b>Important note</b>: the seqname must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.<br><strong>source</strong> - name of the program that generated this feature, or the data source (database or project name)<br><strong>feature</strong> - feature type name, e.g. Gene, Variation, Similarity<br><strong>start</strong> - Start position of the feature, with sequence numbering starting at 1.end - End position of the feature, with sequence numbering starting at 1.<br><strong>score</strong> - A floating point value.strand - defined as + (forward) or - (reverse).<br><strong>frame</strong> - One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on..<br><strong>attribute</strong> - A semicolon-separated list of tag-value pairs, providing additional information about each feature.<br>See more from <i class="fa fa-link"></i><a href="http://www.ensembl.org/info/website/upload/gff.html" target="_blank" rel="external">http://www.ensembl.org/info/website/upload/gff.html</a></p>
<h3 id="genome_files">genome files</h3><p>BEDTools中的一些工具（genomeCoverageBed, complementBed, slopBed）需要物种的染色体大小的信息，genome file一般就是每行都是tab隔开，两列，一列为染色体的名字，第二列为这个染色体的大小。<strong>一般常用物种的genome file在BEDTools安装目录的/genome里面</strong>。<br>自定义基因组genome files文件生成方法见我的另一篇博文：<strong><a href="http://tiramisutes.github.io/2015/11/08/fa-length.html" target="_blank">批量求fasta格式序列长度</a>。</strong></p>
<h2 id="BEDTools使用总结">BEDTools使用总结</h2><h3 id="intersect/intersectBed：计算_Overlaps">intersect/intersectBed：计算 Overlaps</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools intersect <span class="operator">-a</span> A.bed -b B.bed -wa -wb</span><br></pre></td></tr></table></figure>
<p>用来求两个BED或者BAM文件中的overlap，overlap可以进行自定义是整个genome features的overlap还是局部。<br>默认的结果描述如下图：<br><img src="http://7xk19o.com1.z0.glb.clouddn.com/bed1.jpg" alt=""><br>加-wa参数可以报告出原始的在A文件中的feature， 如下图<br><img src="http://7xk19o.com1.z0.glb.clouddn.com/bed2.jpg" alt=""><br>加-wb参数可以报告出原始的在B文件中的feature, 加-c参数可以报告出两个文件中的overlap的feature的数量。<br><i class="fa fa-bolt"></i>当用bedtools intersect 处理大文件时比较耗内存，有效的方法是对A和B文件按照染色体名字(chromosome)和位置(position)排序(<code>sort -k1,1 -k2,2n</code>),然后用<code>-sorted</code>参数重新intersect。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools intersect <span class="operator">-a</span> A-sorted.bed -b B-sorted.bed --sorted</span><br></pre></td></tr></table></figure></p>
<p>其他参数：<br>-wo 返回overlap碱基数<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="variable">$bedtools</span>  intersect <span class="operator">-a</span> A.bed -b B.bed -wo</span><br><span class="line">chr1    <span class="number">0</span>       <span class="number">15</span>      a       chr1    <span class="number">0</span>       <span class="number">4</span>       x       <span class="number">4</span></span><br><span class="line">chr1    <span class="number">0</span>       <span class="number">15</span>      a       chr1    <span class="number">9</span>       <span class="number">15</span>      z       <span class="number">6</span></span><br><span class="line">chr1    <span class="number">25</span>      <span class="number">29</span>      b       chr1    <span class="number">18</span>      <span class="number">28</span>      y       <span class="number">3</span></span><br><span class="line">chr1    <span class="number">18</span>      <span class="number">18</span>      c       chr1    <span class="number">18</span>      <span class="number">28</span>      y       <span class="number">1</span></span><br><span class="line">chr1    <span class="number">10</span>      <span class="number">14</span>      d       chr1    <span class="number">9</span>       <span class="number">15</span>      z       <span class="number">4</span></span><br><span class="line">chr1    <span class="number">20</span>      <span class="number">23</span>      e       chr1    <span class="number">18</span>      <span class="number">28</span>      y       <span class="number">3</span></span><br></pre></td></tr></table></figure></p>
<p>-v 返回非overlap区间<br>-s 相同链上的feature<br>-c 两个文件中的overlap的feature的数量</p>
<h3 id="complement：返回基因组非覆盖区">complement：返回基因组非覆盖区</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools complement -i &lt;BED/GFF/VCF&gt; -g &lt;genome files&gt;</span><br></pre></td></tr></table></figure>
<h3 id="Slop：增加特征区间大小">Slop：增加特征区间大小</h3><p>要求：单个输入bed文件（-i指定）和genome files<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">cat ranges-qry.bed </span><br><span class="line">chr1    <span class="number">0</span>       <span class="number">15</span>      a</span><br><span class="line">chr1    <span class="number">25</span>      <span class="number">29</span>      b</span><br><span class="line">chr1    <span class="number">18</span>      <span class="number">18</span>      c</span><br><span class="line">chr1    <span class="number">10</span>      <span class="number">14</span>      d</span><br><span class="line">chr1    <span class="number">20</span>      <span class="number">23</span>      e</span><br><span class="line">chr1    <span class="number">6</span>       <span class="number">7</span>       f</span><br><span class="line">bedtools slop -i ranges-qry.bed -g genome.txt -b <span class="number">4</span></span><br><span class="line">chr1    <span class="number">0</span>       <span class="number">19</span>      a</span><br><span class="line">chr1    <span class="number">21</span>      <span class="number">33</span>      b</span><br><span class="line">chr1    <span class="number">14</span>      <span class="number">22</span>      c</span><br><span class="line">chr1    <span class="number">6</span>       <span class="number">18</span>      d</span><br><span class="line">chr1    <span class="number">16</span>      <span class="number">27</span>      e</span><br><span class="line">chr1    <span class="number">2</span>       <span class="number">11</span>      f</span><br><span class="line"><span class="comment">#-b 4	:两端同时缩短4个碱基</span></span><br></pre></td></tr></table></figure></p>
<p>-l 3 -r 5:增加左3右5</p>
<h3 id="flank：提取特定区域(启动子区)">flank：提取特定区域(启动子区)</h3><p>要求：基因组GTF文件（-i指定）和genome files<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">bedtools flank -i mm_GRCm38.<span class="number">75</span>_protein_coding_genes.gtf \</span><br><span class="line">                 -g Mus_musculus.GRCm38_genome.txt \</span><br><span class="line">                 <span class="operator">-l</span> <span class="number">3000</span> -r <span class="number">0</span> &gt; mm_GRCm38_3kb_promoters.gtf</span><br><span class="line">cut <span class="operator">-f</span>1,<span class="number">4</span>,<span class="number">5</span>,<span class="number">7</span> mm_GRCm38_3kb_promoters.gtf | head -n <span class="number">3</span></span><br><span class="line"><span class="number">1</span>       <span class="number">3671499</span> <span class="number">3674498</span> -</span><br><span class="line"><span class="number">1</span>       <span class="number">4360315</span> <span class="number">4363314</span> -</span><br><span class="line"><span class="number">1</span>       <span class="number">4496414</span> <span class="number">4499413</span> -</span><br></pre></td></tr></table></figure></p>
<h3 id="getfasta：提取序列">getfasta：提取序列</h3><p>要求：基因组fasta文件（-fi指定）和提取区间GTF文件(-bed指定)<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">bedtools getfasta -fi Mus_musculus.GRCm38.<span class="number">75</span>.dna_rm.toplevel_chr1.fa \</span><br><span class="line">   -bed mm_GRCm38_3kb_promoters.gtf -fo mm_GRCm38_3kb_promoters.fasta</span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-bell"></i>-tab    Report extract sequences in a tab-delimited format instead of in FASTA format.</p>
<p></p><p></p><br><i class="fa fa-commenting-o"></i>提取序列之<code>samtools</code>（速度较快<i class="fa fa-plane"></i>）<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#首先建立fai索引文件（第一列为染色体名字，第二列为序列碱基数）</span></span><br><span class="line">samtools faidx Mus_musculus.GRCm38.<span class="number">75</span>.dna.chromosome.<span class="number">8</span>.fa</span><br><span class="line"><span class="comment">#序列提取，多提取区间空格隔开</span></span><br><span class="line">samtools faidx Mus_musculus.GRCm38.<span class="number">75</span>.dna.chromosome.<span class="number">8</span>.fa \</span><br><span class="line">     <span class="number">8</span>:<span class="number">123407082</span>-<span class="number">123410744</span> <span class="number">8</span>:<span class="number">123518835</span>-<span class="number">123536649</span></span><br><span class="line">&gt;<span class="number">8</span>:<span class="number">123407082</span>-<span class="number">123410744</span></span><br><span class="line">GAGAAAAGCTCCCTTCTTCTCCAGAGTCCCGTCTACCCTGGCTTGGCGAGGGAAAGGAAC</span><br><span class="line">CAGACATATATCAGAGGCAAGTAACCAAGAAGTCTGGAGGTGTTGAGTTTAGGCATGTCT</span><br><span class="line">[...]</span><br><span class="line">&gt;<span class="number">8</span>:<span class="number">123518835</span>-<span class="number">123536649</span></span><br><span class="line">TCTCGCGAGGATTTGAGAACCAGCACGGGATCTAGTCGGAGTTGCCAGGAGACCGCGCAG</span><br><span class="line">CCTCCTCTGACCAGCGCCCATCCCGGATTAGTGGAAGTGCTGGACTGCTGGCACCATGGT</span><br><span class="line">[...]</span><br></pre></td></tr></table></figure><p></p>
<h3 id="nuc:_计算GC含量即各碱基数">nuc: 计算GC含量即各碱基数</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bedtools nuc -fi hg19.fa -bed CDS.bed</span><br></pre></td></tr></table></figure>
<p>输出结果解释：在原bed文件每行结尾增加以下几列<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">Output format: </span><br><span class="line">The following information will be reported after each BED entry:</span><br><span class="line">    <span class="number">1</span>) %AT content</span><br><span class="line">    <span class="number">2</span>) %GC content</span><br><span class="line">    <span class="number">3</span>) Number of As observed</span><br><span class="line">    <span class="number">4</span>) Number of Cs observed</span><br><span class="line">    <span class="number">5</span>) Number of Gs observed</span><br><span class="line">    <span class="number">6</span>) Number of Ts observed</span><br><span class="line">    <span class="number">7</span>) Number of Ns observed</span><br><span class="line">    <span class="number">8</span>) Number of other bases observed</span><br><span class="line">    <span class="number">9</span>) The length of the explored sequence/interval.</span><br><span class="line">    <span class="number">10</span>) The seq. extracted from the FASTA file. (opt., <span class="keyword">if</span> -seq is used)</span><br><span class="line">    <span class="number">11</span>) The number of <span class="built_in">times</span> a user<span class="string">'s pattern was observed.</span><br><span class="line">        (opt., if -pattern is used.)</span></span><br></pre></td></tr></table></figure></p>
<h3 id="genomecov：染色体和全基因组覆盖度计算">genomecov：染色体和全基因组覆盖度计算</h3><p>要求：单个输入bed文件（-i指定）和genome files；如果输入为bam(-ibam指定)文件，则不需要genome files。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">cat ranges-cov-sorted.bed</span><br><span class="line">chr1    <span class="number">4</span>       <span class="number">9</span></span><br><span class="line">chr1    <span class="number">1</span>       <span class="number">6</span></span><br><span class="line">chr1    <span class="number">8</span>       <span class="number">19</span></span><br><span class="line">chr1    <span class="number">25</span>      <span class="number">30</span></span><br><span class="line">chr2    <span class="number">0</span>       <span class="number">20</span></span><br><span class="line">$ cat cov.txt</span><br><span class="line">chr1    <span class="number">30</span></span><br><span class="line">chr2    <span class="number">20</span></span><br><span class="line">bedtools genomecov -i ranges-cov-sorted.bed -g cov.txt</span><br><span class="line">chr1    <span class="number">0</span>       <span class="number">7</span>       <span class="number">30</span>      <span class="number">0.233333</span> <span class="number">1</span></span><br><span class="line">chr1    <span class="number">1</span>       <span class="number">20</span>      <span class="number">30</span>      <span class="number">0.666667</span></span><br><span class="line">chr1    <span class="number">2</span>       <span class="number">3</span>       <span class="number">30</span>      <span class="number">0.1</span></span><br><span class="line">chr2    <span class="number">1</span>       <span class="number">20</span>      <span class="number">20</span>      <span class="number">1</span> <span class="number">2</span></span><br><span class="line">genome  <span class="number">0</span>       <span class="number">7</span>       <span class="number">50</span>      <span class="number">0.14</span> <span class="number">3</span></span><br><span class="line">genome  <span class="number">1</span>       <span class="number">40</span>      <span class="number">50</span>      <span class="number">0.8</span></span><br><span class="line">genome  <span class="number">2</span>       <span class="number">3</span>       <span class="number">50</span>      <span class="number">0.06</span></span><br><span class="line"><span class="comment">#name 覆盖次数 覆盖碱基数 总碱基数   覆盖度</span></span><br><span class="line"><span class="comment">#同时计算单染色体和全基因组覆盖度</span></span><br></pre></td></tr></table></figure></p>
<p><i class="fa fa-cogs"></i><li>ranges-cov.bed文件需提前排序sort -k1,1 ranges-cov.bed &gt; ranges-cov-sorted.bed</li></p>
<p><li>-bg参数可得到每个碱基的覆盖度。</li></p>
<h3 id="coverage：计算染色体给定区间覆盖度">coverage：计算染色体给定区间覆盖度</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">$ cat A.bed</span><br><span class="line">chr1  <span class="number">0</span>   <span class="number">100</span></span><br><span class="line">chr1  <span class="number">100</span> <span class="number">200</span></span><br><span class="line">chr2  <span class="number">0</span>   <span class="number">100</span></span><br><span class="line"></span><br><span class="line">$ cat B.bed</span><br><span class="line">chr1  <span class="number">10</span>  <span class="number">20</span></span><br><span class="line">chr1  <span class="number">20</span>  <span class="number">30</span></span><br><span class="line">chr1  <span class="number">30</span>  <span class="number">40</span></span><br><span class="line">chr1  <span class="number">100</span> <span class="number">200</span></span><br><span class="line"></span><br><span class="line">$ bedtools coverage <span class="operator">-a</span> A.bed -b B.bed</span><br><span class="line">chr1  <span class="number">0</span>   <span class="number">100</span>  <span class="number">3</span>  <span class="number">30</span>  <span class="number">100</span> <span class="number">0.3000000</span></span><br><span class="line">chr1  <span class="number">100</span> <span class="number">200</span>  <span class="number">1</span>  <span class="number">100</span> <span class="number">100</span> <span class="number">1.0000000</span></span><br><span class="line">chr2  <span class="number">0</span>   <span class="number">100</span>  <span class="number">0</span>  <span class="number">0</span>   <span class="number">100</span> <span class="number">0.0000000</span></span><br></pre></td></tr></table></figure>
<p></p><p></p><br>贡献来源<br><a href="http://www.plob.org/2012/09/26/3748.html" target="_blank" rel="external">http://www.plob.org/2012/09/26/3748.html</a><br><a href="http://bedtools.readthedocs.org/en/latest/content/bedtools-suite.html" target="_blank" rel="external">http://bedtools.readthedocs.org/en/latest/content/bedtools-suite.html</a><br><a href="https://code.google.com/archive/p/bedtools/wikis/Usage.wiki" target="_blank" rel="external">https://code.google.com/archive/p/bedtools/wikis/Usage.wiki</a><br><a href="https://code.google.com/archive/p/bedtools/wikis/UsageAdvanced.wiki" target="_blank" rel="external">https://code.google.com/archive/p/bedtools/wikis/UsageAdvanced.wiki</a><p></p>

      
    </div>

    <div>
      
        
<div id="wechat_subscriber" style="display: block; padding: 10px 0; margin: 20px auto; width: 100%; text-align: center">
    <img id="wechat_subscriber_qcode" src="/images/Wechat.jpg" alt="tiramisutes wechat" style="width: 200px; max-width: 100%;"/>
    <div>欢迎关注</div>
</div>


      
    </div>

    <div>
      
        

      
    </div>

    <div>
      
        
  <ul class="post-copyright">
    <li class="post-copyright-author">
      <strong>本文作者:</strong>
      tiramisutes
    </li>
    <li class="post-copyright-link">
      <strong>本文固定链接:</strong>
      <a href="http://tiramisutes.github.io/2016/03/18/bedtools.html" title="bedtools 使用小结">http://tiramisutes.github.io/2016/03/18/bedtools.html</a>
      <strong>转载请注明出处！</strong>
    </li>
    <li class="post-copyright-license">
      <strong>Copyright Notice: </strong>
      All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/3.0/" rel="external nofollow" target="_blank">CC BY-NC-SA 3.0</a> unless stating additionally.
    </li>
  </ul>


      
    </div>

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/BEDTools/" rel="tag"># BEDTools</a>
          
        </div>
      

      
        
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2016/03/15/dplyr.html" rel="next" title="R数据整形术之 dplyr">
                <i class="fa fa-chevron-left"></i> R数据整形术之 dplyr
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2016/03/20/Samtools.html" rel="prev" title="被忽视的Samtools参数">
                被忽视的Samtools参数 <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </article>



    <div class="post-spread">
      
        <!-- JiaThis Button BEGIN -->
<div class="jiathis_style">
  <a class="jiathis_button_tsina"></a>
  <a class="jiathis_button_tqq"></a>
  <a class="jiathis_button_weixin"></a>
  <a class="jiathis_button_cqq"></a>
  <a class="jiathis_button_douban"></a>
  <a class="jiathis_button_renren"></a>
  <a class="jiathis_button_qzone"></a>
  <a class="jiathis_button_kaixin001"></a>
  <a class="jiathis_button_copy"></a>
  <a href="http://www.jiathis.com/share" class="jiathis jiathis_txt jiathis_separator jtico jtico_jiathis" target="_blank"></a>
  <a class="jiathis_counter_style"></a>
</div>
<script type="text/javascript" >
  var jiathis_config={
    hideMore:false
  }
</script>
<script type="text/javascript" src="http://v3.jiathis.com/code/jia.js" charset="utf-8"></script>
<!-- JiaThis Button END -->

      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
      <div id="hypercomments_widget"></div>
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            Overview
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/images/author.jpg"
               alt="tiramisutes" />
          <p class="site-author-name" itemprop="name">tiramisutes</p>
           
              <p class="site-description motion-element" itemprop="description"></p>
           
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">129</span>
                <span class="site-state-item-name">posts</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">11</span>
                <span class="site-state-item-name">categories</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">100</span>
                <span class="site-state-item-name">tags</span>
              </a>
            </div>
          

        </nav>

        
          <div class="feed-link motion-element">
            <a href="/atom.xml" rel="alternate">
              <i class="fa fa-rss"></i>
              RSS
            </a>
          </div>
        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="https://github.com/tiramisutes" target="_blank" title="GitHub">
                  
                    <i class="fa fa-fw fa-github"></i>
                  
                  GitHub
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="https://twitter.com/xzpgocxx" target="_blank" title="Twitter">
                  
                    <i class="fa fa-fw fa-twitter"></i>
                  
                  Twitter
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://weibo.com/tiramisutesTes" target="_blank" title="Weibo">
                  
                    <i class="fa fa-fw fa-weibo"></i>
                  
                  Weibo
                </a>
              </span>
            
          
        </div>

        
        

        
        
          <div class="links-of-blogroll motion-element links-of-blogroll-inline">
            <div class="links-of-blogroll-title">
              <i class="fa  fa-fw fa-globe"></i>
              Links
            </div>
            <ul class="links-of-blogroll-list">
              
                <li class="links-of-blogroll-item">
                  <a href="https://tiramisutes.gitbooks.io/awk/content/" title="GitBook《awk学习总结》" target="_blank">GitBook《awk学习总结》</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://cotton.hzau.edu.cn/" title="棉花遗传改良团队" target="_blank">棉花遗传改良团队</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.bio-info-trainee.com/" title="生信菜鸟团-深圳大学城" target="_blank">生信菜鸟团-深圳大学城</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.biotrainee.com/forum.php" title="生信技能树" target="_blank">生信技能树</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.iwhgao.com/" title="rabbit gao's blog" target="_blank">rabbit gao's blog</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://huans.github.io/" title="Huans(代谢+GWAS)" target="_blank">Huans(代谢+GWAS)</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.sequencing.uio.no/" title="NSC-sequenceing" target="_blank">NSC-sequenceing</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://blog.jobbole.com/" title="伯乐在线" target="_blank">伯乐在线</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://linux.cn/" title="Linux开源中文社区" target="_blank">Linux开源中文社区</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.r-bloggers.com/" title="R-bloggers" target="_blank">R-bloggers</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://cran.r-project.org/other-docs.html" title="R-resource" target="_blank">R-resource</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html" title="ggplot2" target="_blank">ggplot2</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://lilibei.net/" title="pele' blog" target="_blank">pele' blog</a>
                </li>
              
            </ul>
          </div>
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#概述"><span class="nav-number">1.</span> <span class="nav-text">概述</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#与BEDTools使用相关的基本概念"><span class="nav-number">2.</span> <span class="nav-text">与BEDTools使用相关的基本概念</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#BED和GFF文件的一个差异"><span class="nav-number">3.</span> <span class="nav-text">BED和GFF文件的一个差异</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#相关格式"><span class="nav-number">4.</span> <span class="nav-text">相关格式</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#BED_format"><span class="nav-number">4.1.</span> <span class="nav-text">BED format</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#GFF_format"><span class="nav-number">4.2.</span> <span class="nav-text">GFF format</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#genome_files"><span class="nav-number">4.3.</span> <span class="nav-text">genome files</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#BEDTools使用总结"><span class="nav-number">5.</span> <span class="nav-text">BEDTools使用总结</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#intersect/intersectBed：计算_Overlaps"><span class="nav-number">5.1.</span> <span class="nav-text">intersect/intersectBed：计算 Overlaps</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#complement：返回基因组非覆盖区"><span class="nav-number">5.2.</span> <span class="nav-text">complement：返回基因组非覆盖区</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Slop：增加特征区间大小"><span class="nav-number">5.3.</span> <span class="nav-text">Slop：增加特征区间大小</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#flank：提取特定区域(启动子区)"><span class="nav-number">5.4.</span> <span class="nav-text">flank：提取特定区域(启动子区)</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#getfasta：提取序列"><span class="nav-number">5.5.</span> <span class="nav-text">getfasta：提取序列</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#nuc:_计算GC含量即各碱基数"><span class="nav-number">5.6.</span> <span class="nav-text">nuc: 计算GC含量即各碱基数</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#genomecov：染色体和全基因组覆盖度计算"><span class="nav-number">5.7.</span> <span class="nav-text">genomecov：染色体和全基因组覆盖度计算</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#coverage：计算染色体给定区间覆盖度"><span class="nav-number">5.8.</span> <span class="nav-text">coverage：计算染色体给定区间覆盖度</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy;  2014 - 
  <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder"> <a href="/about" target="_blank">tiramisutes</a></span>
</div>

<div class="powered-by">
  Powered by <a class="theme-link" href="https://hexo.io" target="_blank">Hexo</a>
</div>

<div class="theme-info">
  Theme -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next" target="_blank">
    NexT.Muse 
  </a>
<script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js">
                </script>
    <span id="busuanzi_container_site_pv" style='display:none'>
                      &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;<a style="line-height: 1.5;" href="http://c4.gostats.cn/last_guests.xml?id=397056" target="_blank">The total visits <span id="busuanzi_value_site_pv"></span> times</a>
                </span>
    <!-- GoStats JavaScript Based Code -->
                <script type="text/javascript" src="http://gostats.cn/js/counter.js"></script>
                <script type="text/javascript">_gos='c4.gostats.cn';_goa=397056;
                _got=5;_goi=1;_GoStatsRun();</script>
                <noscript><a target="_blank" 
                href="http://gostats.cn"><img src="http://c4.gostats.cn/bin/count/a_397056/t_5/i_1/counter.png" 
                style="border-width:0" /></a></noscript>
                <!-- End GoStats JavaScript Based Code -->  
</div>
<script>
 host3 = window.location.href;
  chinadomai3 = "http://tiramisutes.github.io/";
  if (host3.indexOf(chinadomain3)!=-1)  {

  (function(){
    var bp = document.createElement('script');
    bp.src = '//push.zhanzhang.baidu.com/push.js';
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>
        

        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.1"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.1"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.1"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.1"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.1"></script>



  


  




	

		<script type="text/javascript">
		_hcwp = window._hcwp || [];

		_hcwp.push({widget:"Bloggerstream", widget_id: 91479, selector:".hc-comment-count", label: "{\%COUNT%\}" });

		
		_hcwp.push({widget:"Stream", widget_id: 91479, xid: "2016/03/18/bedtools.html"});
		

		(function() {
		if("HC_LOAD_INIT" in window)return;
		HC_LOAD_INIT = true;
		var lang = (navigator.language || navigator.systemLanguage || navigator.userLanguage || "en").substr(0, 2).toLowerCase();
		var hcc = document.createElement("script"); hcc.type = "text/javascript"; hcc.async = true;
		hcc.src = ("https:" == document.location.protocol ? "https" : "http")+"://w.hypercomments.com/widget/hc/91479/"+lang+"/widget.js";
		var s = document.getElementsByTagName("script")[0];
		s.parentNode.insertBefore(hcc, s.nextSibling);
		})();
		</script>

	












  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (search_path.endsWith("json")) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.1.js"></script>
  <script>AV.initialize("c7x0fPmO8Ci9avDBRuGWHqpG-gzGzoHsz", "iVJ6ST0e1O5xaiIhVwOyhbf3");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  
<script>
(function(){
    var bp = document.createElement('script');
    var curProtocol = window.location.protocol.split(':')[0];
    if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
    }
    else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
    }
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


  
  


  
  <script type="text/javascript" src="/js/src/js.cookie.js?v=5.1.1"></script>
  <script type="text/javascript" src="/js/src/scroll-cookie.js?v=5.1.1"></script>


  

</body>
</html>
