<!doctype html>



  


<html class="theme-next muse use-motion" lang="default">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>



  
  
  <link rel="stylesheet" media="all" href="/lib/Han/dist/han.min.css?v=3.3">




<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="b12Zu40L63PLgCtcn0IW8KCNRolmyQlaF22tf4rRLEg" />













  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="http://fonts.useso.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.1" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="Samtools," />





  <link rel="alternate" href="/atom.xml" title="hope" type="application/atom+xml" />




  <link rel="shortcut icon" type="image/x-icon" href="/images/favicon.ico?v=5.1.1" />






<meta name="description" content="Samtools是一个用于操作序列比对结果sam和bam文件的工具合集。
sam文件格式SAM格式由两部分组成：头部区和比对区，都以tab分列。头部区:以’@’开始，体现了比对的一些总体信息。比对的SAM格式版本，比对的参考序列，比对使用的软件等。比对区: 比对结果，每一个比对结果是一行，有11个主列和1个可选列。123456789101112131415161718192021222324252">
<meta property="og:type" content="article">
<meta property="og:title" content="被忽视的Samtools参数">
<meta property="og:url" content="http://tiramisutes.github.io/2016/03/20/Samtools.html">
<meta property="og:site_name" content="hope">
<meta property="og:description" content="Samtools是一个用于操作序列比对结果sam和bam文件的工具合集。
sam文件格式SAM格式由两部分组成：头部区和比对区，都以tab分列。头部区:以’@’开始，体现了比对的一些总体信息。比对的SAM格式版本，比对的参考序列，比对使用的软件等。比对区: 比对结果，每一个比对结果是一行，有11个主列和1个可选列。123456789101112131415161718192021222324252">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/flag.png">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/sam_output2.png">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/cigar.png">
<meta property="og:updated_time" content="2017-09-11T13:23:05.198Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="被忽视的Samtools参数">
<meta name="twitter:description" content="Samtools是一个用于操作序列比对结果sam和bam文件的工具合集。
sam文件格式SAM格式由两部分组成：头部区和比对区，都以tab分列。头部区:以’@’开始，体现了比对的一些总体信息。比对的SAM格式版本，比对的参考序列，比对使用的软件等。比对区: 比对结果，每一个比对结果是一行，有11个主列和1个可选列。123456789101112131415161718192021222324252">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":0,"b2t":false,"scrollpercent":false},
    fancybox: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://tiramisutes.github.io/2016/03/20/Samtools.html"/>





  <title>被忽视的Samtools参数 | hope</title>
</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?08b17c69c7ba460d9ab5f188a048ee6d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>











  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">hope</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <h1 class="site-subtitle" itemprop="description">A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</h1>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      
        
        <li class="menu-item menu-item-code">
          <a href="/code" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-code"></i> <br />
            
            Code
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-footprint">
          <a href="/footprint" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-image"></i> <br />
            
            Footprint
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            Search
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal " itemscope itemtype="http://schema.org/Article">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2016/03/20/Samtools.html">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/author.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope">
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">被忽视的Samtools参数</h2>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2016-03-20T22:54:42+08:00">
                2016-03-20
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index">
                    <span itemprop="name">Bioinformatics</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
            <!--noindex-->
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2016/03/20/Samtools.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count hc-comment-count" data-xid="2016/03/20/Samtools.html" itemprop="commentsCount"></span>
                </a>
              </span>
              <!--/noindex-->
            
          

          
          
             <span id="/2016/03/20/Samtools.html" class="leancloud_visitors" data-flag-title="被忽视的Samtools参数">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Visitors </span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-wordcount">
              
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article</span>
                
                <span title="Words count in article">
                  3,500
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time</span>
                
                <span title="Reading time">
                  19
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        <p><a href="http://www.plob.org/tag/samtools" title="View all posts in samtools" target="_blank">Samtools</a>是一个用于操作序列比对结果sam和bam文件的工具合集。</p>
<h2 id="sam文件格式">sam文件格式</h2><p>SAM格式由两部分组成：头部区和比对区，都以tab分列。<br><strong>头部区:</strong>以’@’开始，体现了比对的一些总体信息。比对的SAM格式版本，比对的参考序列，比对使用的软件等。<br><strong>比对区:</strong> 比对结果，每一个比对结果是一行，有11个主列和1个可选列。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line">@HD VN:<span class="number">1.0</span> SO:unsorted  </span><br><span class="line">头部区第一行：VN是格式版本；SO表示比对排序的类型，有unkown(default),unsorted,</span><br><span class="line">queryname和coordinate几种。samtools软件在进行排序后不能自动更新bam文件的SO值。</span><br><span class="line">picard却可以。</span><br><span class="line">@SQ SN:A.auricula_all_contig_1 LN:<span class="number">9401</span></span><br><span class="line">参考序列名。这些参考序列决定了比对结果sort的顺序。SN是参考序列名；LN是参考序列</span><br><span class="line">长度;</span><br><span class="line">@RG ID:sample01</span><br><span class="line">Read Group. <span class="number">1</span>个sample的测序结果为<span class="number">1</span>个Read Group；该sample可以有多个library</span><br><span class="line">的测序结果。</span><br><span class="line">@PG ID:bowtie2 PN:bowtie2 VN:<span class="number">2.0</span>.<span class="number">0</span>-beta7</span><br><span class="line">比对所使用的软件。</span><br><span class="line"></span><br><span class="line">比对区<span class="number">11</span>个列和可选列的解释</span><br><span class="line"><span class="number">1</span>  QNAME  比对的序列名,即单端或双端fa/fq 中的reads编号</span><br><span class="line"><span class="number">2</span>  FLAG   Bwise FLAG(表明比对类型：pairing，strand，mate strand等)</span><br><span class="line"><span class="number">3</span>  RNAME  比对上的参考序列名</span><br><span class="line"><span class="number">4</span>  POS    <span class="number">1</span>-Based的比对上的最左边的定位</span><br><span class="line"><span class="number">5</span>  MAPQ   比对质量，<span class="number">255</span>表示没有map</span><br><span class="line"><span class="number">6</span>  CIGAR  Extended CIGAR string (操作符：MIDNSHP) 比对结果信息：匹配碱基数，可变剪接等，*表示不可用。</span><br><span class="line"><span class="number">7</span>  MRNM   相匹配的另外一条序列，比对上的参考序列名</span><br><span class="line"><span class="number">8</span>  MPOS   <span class="number">1</span>-Based leftmost Mate POsition</span><br><span class="line"><span class="number">9</span>  ISIZE  插入片段长度</span><br><span class="line"><span class="number">10</span> SEQ    和参考序列在同一个琏上的比对序列(若比对结果在负意链上，则序列是其反向重复序列)</span><br><span class="line"><span class="number">11</span> QUAL   比对序列的质量(ASCII-<span class="number">33</span>=Phred base quality)</span><br><span class="line"><span class="number">12</span> 可选的行，以TAG：TYPE：VALUE的形式提供额外的信息</span><br></pre></td></tr></table></figure></p>
<h3 id="比对区解释">比对区解释</h3><p><strong>sam/bam比对区包含有此次比对的结果信息，</strong>其中主要信息解释如下：</p>
<p><li><strong>FLAG部分</strong></li><br><img src="http://7xk19o.com1.z0.glb.clouddn.com/flag.png" alt=""><br><strong>0x800</strong> 表明相应位置的比对属于嵌合体比对；<br><strong>0x4</strong> 没有map上的reads；<br><img src="http://7xk19o.com1.z0.glb.clouddn.com/sam_output2.png" alt=""></p>
<p><li><strong>CIGAR部分</strong></li><br><img src="http://7xk19o.com1.z0.glb.clouddn.com/cigar.png" alt=""><br>对于mRNA到基因组的比对，N表示内含子。<br>More: <a href="http://samtools.github.io/hts-specs/SAMv1.pdf" target="_blank" rel="external">http://samtools.github.io/hts-specs/SAMv1.pdf</a></p>
<h2 id="sam文件的几个特例解释">sam文件的几个特例解释</h2><h3 id="Unmapped_reads_统计">Unmapped reads 统计</h3><p>Each alignment is one line of the SAM file, but not all lines are successful alignments. <strong>Unmapped reads在sam文件中的标记：<code>FLAG</code>列为4而且<code>RNAME</code>列为星号*；</strong><br><figure class="highlight nimrod"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#统计包含星号的比对行数</span></span><br><span class="line">cut -f3 smallRNA-<span class="type">seq</span>.sam | grep -c \*</span><br><span class="line"><span class="comment">#总的比对行数</span></span><br><span class="line">grep -c -v <span class="string">"^@"</span> smallRNA-<span class="type">seq</span>.sam</span><br></pre></td></tr></table></figure></p>
<h3 id="How_many_different_read_IDs_are_in_the_file?">How many different read IDs are in the file?</h3><p>The query (read) ID is in field 1. Some reads may have multiple alignments, so the number of lines is not necessarily the number of reads.<br><figure class="highlight gherkin"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">grep -v <span class="string">"@"</span> HR-1B.fq.gz.sam |<span class="string"> cut -f1 </span>|<span class="string"> sort </span>|<span class="string"> uniq </span>|<span class="string"> wc -l </span><br><span class="line">#同时统计此次比对的单端/双端 fa/fq文件发现两者结果不同，说明确实有些reads未能比对上去。</span><br><span class="line">zcat pepper/RNA-seq/sgs-clean-reads/HR-1B.fq.gz </span>|<span class="string"> wc -l</span></span><br></pre></td></tr></table></figure></p>
<h3 id="How_many_different_read_sequences_are_in_the_file?">How many different read sequences are in the file?</h3><p>第一列的reads ID仅能表示测序过程中的不同reads，但他们的序列可能因为PCR扩增原因或文库偏好而完全相同，所以统计第10列的uniq序列数能够准确的表示reads总数。<br><figure class="highlight gherkin"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cut -f10 smallRNAseq.sam |<span class="string"> sort </span>|<span class="string"> uniq </span>|<span class="string"> wc -l</span></span><br></pre></td></tr></table></figure></p>
<h3 id="How_many_reads_are_uniquely_mapped?">How many reads are <strong>uniquely mapped</strong>?</h3><figure class="highlight gherkin"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cut -f10 smallRNAseq.sam |<span class="string"> sort </span>|<span class="string"> uniq -u </span>|<span class="string"> wc -l</span></span><br></pre></td></tr></table></figure>
<p>对于 BWA比对结果也可用<code>grep -c XT:A:U smallRNA-seq.sam</code>来准确统计。</p>
<h3 id="How_many_reads_are_multi-hits?">How many reads are <strong>multi-hits</strong>?</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cut <span class="operator">-f</span>1 smallRNA-seq.sam | sort | uniq <span class="operator">-d</span> | wc <span class="operator">-l</span></span><br></pre></td></tr></table></figure>
<p>对于 BWA比对结果也可用<code>grep -c XT:A:R smallRNAseq.sam</code>来准确统计。</p>
<h3 id="How_many_alignments_are_reported_for_each_read?">How many alignments are reported for each read?</h3><figure class="highlight gherkin"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">grep -v <span class="string">"^@"</span> smallRNA-seq.sam |<span class="string"> cut -f1 </span>|<span class="string"> sort </span>|<span class="string"> uniq -c </span>|<span class="string"> sort -nr &gt; sortedreadcount.txt</span><br><span class="line">grep -v "^@" smallRNA-seq.sam </span>|<span class="string"> cut -f1 </span>|<span class="string">sort </span>|<span class="string"> uniq -c </span>|<span class="string"> sort -nr </span>|<span class="string"> cut -c1-8 </span>|<span class="string"> sort </span>|<span class="string"> uniq -c</span></span><br></pre></td></tr></table></figure>
<h3 id="How_many_different_reference_sequences_are_represented_in_the_file?">How many different reference sequences are represented in the file?</h3><figure class="highlight gherkin"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">grep -v <span class="string">"^@"</span> smallRNA-seq.sam |<span class="string"> cut –f3 </span>|<span class="string"> sort </span>|<span class="string"> uniq </span>|<span class="string"> wc -l</span></span><br></pre></td></tr></table></figure>
<h2 id="view">view</h2><p>-c    计数<br>-f    返回指定区间/flags比对结果<br>-q    返回比对质量大于等于指定值的比对数目<br><strong>-F 4：</strong>统计map 上的 reads总数；<br><strong>-f 4：</strong>统计没有map 上的 reads总数；<br>To get the <strong>unmapped</strong> reads from a <a rel="external" href="http://samtools.sourceforge.net/SAM1.pdf" target="_blank">bam</a> file use :<br><code>samtools view -f 4 file.bam &gt; unmapped.sam</code>, the output will be in <a rel="external" href="http://samtools.sourceforge.net/SAM1.pdf" target="_blank">sam</a><br>to get the output in <a rel="external" href="http://samtools.sourceforge.net/SAM1.pdf" target="_blank">bam</a> use : <code>samtools view -b -f 4 file.bam &gt; unmapped.bam</code><br>To get only the <strong>mapped</strong> reads use the parameter ‘F’, which works like <code>-v</code> of grep and skips the alignments for a specific flag.<br><code>samtools view -b -F 4 file.bam &gt; mapped.bam</code><br>samtools view -b -F 4 -f 8 file.bam &gt; onlyThisEndMapped.bam<br>samtools view -b -F 8 -f 4 file.bam &gt; onlyThatEndMapped.bam<br>samtools view -b -F12 file.bam &gt; bothEndsMapped.bam<br>samtools merge merged.bam onlyThisEndMapped.bam onlyThatEndMapped.bam bothEndsMapped.bam<br><strong>对于tophat比对结果：</strong><br><code><span style="background-color: rgb(249, 242, 244);">samtools view -b -f 2  accepted_hits.bam &gt; mappedPairs.bam</span></code><br>Better with:<br><code><span style="background-color: rgb(249, 242, 244);">samtools view -b -f 0x2 accepted_hits.bam &gt; mappedPairs.bam</span></code></p>
<h2 id="sort">sort</h2><p>-m    指定运算内存，支持K，M，G等缩写<br>-@    并行运算核数</p>
<h2 id="index">index</h2><p>必须对bam文件进行默认情况下的排序后，才能进行index。否则会报错。</p>
<p>建立索引后将产生后缀为.bai的文件，用于快速的随机处理。很多情况下需要有bai文件的存在，特别是显示序列比对情况下。比如samtool的tview命令就需要；gbrowse2显示reads的比对图形的时候也需要。</p>
<h2 id="faidx">faidx</h2><p>对fasta文件建立索引,生成的索引文件以.fai后缀结尾。该命令也能依据索引文件快速提取fasta文件中的某一条（子）序列。<br>See more： <a href="http://tiramisutes.github.io/2016/03/18/bedtools.html" target="_blank">bedtools 使用小结</a></p>
<h2 id="flagstat">flagstat</h2><p>给出BAM文件的比对结果<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">$ samtools flagstat example.bam</span><br><span class="line"><span class="number">11945742</span> + <span class="number">0</span> <span class="keyword">in</span> total (QC-passed reads + QC-failed reads)</span><br><span class="line"><span class="comment">#总共的reads数</span></span><br><span class="line"><span class="number">0</span> + <span class="number">0</span> duplicates</span><br><span class="line"><span class="number">7536364</span> + <span class="number">0</span> mapped (<span class="number">63.09</span>%:-nan%)</span><br><span class="line"><span class="comment">#总体上reads的匹配率</span></span><br><span class="line"><span class="number">11945742</span> + <span class="number">0</span> paired <span class="keyword">in</span> sequencing</span><br><span class="line"><span class="comment">#有多少reads是属于paired reads</span></span><br><span class="line"><span class="number">5972871</span> + <span class="number">0</span> <span class="built_in">read</span>1</span><br><span class="line"><span class="comment">#reads1中的reads数</span></span><br><span class="line"><span class="number">5972871</span> + <span class="number">0</span> <span class="built_in">read</span>2</span><br><span class="line"><span class="comment">#reads2中的reads数</span></span><br><span class="line"><span class="number">6412042</span> + <span class="number">0</span> properly paired (<span class="number">53.68</span>%:-nan%)</span><br><span class="line"><span class="comment">#完美匹配的reads数：比对到同一条参考序列，并且两条reads之间的距离符合设置的阈值</span></span><br><span class="line"><span class="number">6899708</span> + <span class="number">0</span> with itself and mate mapped</span><br><span class="line"><span class="comment">#paired reads中两条都比对到参考序列上的reads数</span></span><br><span class="line"><span class="number">636656</span> + <span class="number">0</span> singletons (<span class="number">5.33</span>%:-nan%)</span><br><span class="line"><span class="comment">#单独一条匹配到参考序列上的reads数，和上一个相加，则是总的匹配上的reads数。</span></span><br><span class="line"><span class="number">469868</span> + <span class="number">0</span> with mate mapped to a different chr</span><br><span class="line"><span class="comment">#paired reads中两条分别比对到两条不同的参考序列的reads数</span></span><br><span class="line"><span class="number">243047</span> + <span class="number">0</span> with mate mapped to a different chr (mapQ&gt;=<span class="number">5</span>)</span><br></pre></td></tr></table></figure></p>
<h2 id="mpileup">mpileup</h2><p>samtools还有个非常重要的命令mpileup，以前为pileup。该命令用于生成bcf文件，再使用bcftools进行SNP和Indel的分析。bcftools是samtool中附带的软件，在samtools的安装文件夹中可以找到。<br>-f 来输入有索引文件的fasta参考序列；<br>-g 输出到bcf格式。</p>
<h2 id="depth">depth</h2><p>计算每一个位点或者区域的测序深度；<br><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools depth sorted<span class="class">.bam</span> &gt; sorted<span class="class">.bam</span><span class="class">.txt</span></span><br></pre></td></tr></table></figure></p>
<p>一共得到3列以指标分隔符分隔的数据，第一列为染色体名称，第二列为位点，第三列为覆盖深度。</p>
<p></p><p></p><br>贡献来源<br><a href="http://www.plob.org/2014/01/26/7112.html" target="_blank" rel="external">http://www.plob.org/2014/01/26/7112.html</a><br><a href="http://blog.sina.com.cn/s/blog_670445240101l30k.html" target="_blank" rel="external">http://blog.sina.com.cn/s/blog_670445240101l30k.html</a><br><a href="https://www.biostars.org/p/56246/" target="_blank" rel="external">https://www.biostars.org/p/56246/</a><br><a href="https://www.biostars.org/p/110039/" target="_blank" rel="external">https://www.biostars.org/p/110039/</a><br><a href="https://www.biostars.org/p/95929/" target="_blank" rel="external">https://www.biostars.org/p/95929/</a><br><a href="https://www.biostars.org/p/110157/" target="_blank" rel="external">https://www.biostars.org/p/110157/</a><br><a href="https://groups.google.com/forum/#!forum/bedtools-discuss" target="_blank" rel="external">https://groups.google.com/forum/#!forum/bedtools-discuss</a><br><a href="https://code.google.com/p/hydra-sv/" target="_blank" rel="external">https://code.google.com/p/hydra-sv/</a><p></p>

      
    </div>

    <div>
      
        
<div id="wechat_subscriber" style="display: block; padding: 10px 0; margin: 20px auto; width: 100%; text-align: center">
    <img id="wechat_subscriber_qcode" src="/images/Wechat.jpg" alt="tiramisutes wechat" style="width: 200px; max-width: 100%;"/>
    <div>欢迎关注</div>
</div>


      
    </div>

    <div>
      
        

      
    </div>

    <div>
      
        
  <ul class="post-copyright">
    <li class="post-copyright-author">
      <strong>本文作者:</strong>
      tiramisutes
    </li>
    <li class="post-copyright-link">
      <strong>本文固定链接:</strong>
      <a href="http://tiramisutes.github.io/2016/03/20/Samtools.html" title="被忽视的Samtools参数">http://tiramisutes.github.io/2016/03/20/Samtools.html</a>
      <strong>转载请注明出处！</strong>
    </li>
    <li class="post-copyright-license">
      <strong>Copyright Notice: </strong>
      All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/3.0/" rel="external nofollow" target="_blank">CC BY-NC-SA 3.0</a> unless stating additionally.
    </li>
  </ul>


      
    </div>

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/Samtools/" rel="tag"># Samtools</a>
          
        </div>
      

      
        
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2016/03/18/bedtools.html" rel="next" title="bedtools 使用小结">
                <i class="fa fa-chevron-left"></i> bedtools 使用小结
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2016/03/21/find-rm.html" rel="prev" title="Linux中如何正确删除：find-rm">
                Linux中如何正确删除：find-rm <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </article>



    <div class="post-spread">
      
        <!-- JiaThis Button BEGIN -->
<div class="jiathis_style">
  <a class="jiathis_button_tsina"></a>
  <a class="jiathis_button_tqq"></a>
  <a class="jiathis_button_weixin"></a>
  <a class="jiathis_button_cqq"></a>
  <a class="jiathis_button_douban"></a>
  <a class="jiathis_button_renren"></a>
  <a class="jiathis_button_qzone"></a>
  <a class="jiathis_button_kaixin001"></a>
  <a class="jiathis_button_copy"></a>
  <a href="http://www.jiathis.com/share" class="jiathis jiathis_txt jiathis_separator jtico jtico_jiathis" target="_blank"></a>
  <a class="jiathis_counter_style"></a>
</div>
<script type="text/javascript" >
  var jiathis_config={
    hideMore:false
  }
</script>
<script type="text/javascript" src="http://v3.jiathis.com/code/jia.js" charset="utf-8"></script>
<!-- JiaThis Button END -->

      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
      <div id="hypercomments_widget"></div>
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            Overview
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/images/author.jpg"
               alt="tiramisutes" />
          <p class="site-author-name" itemprop="name">tiramisutes</p>
           
              <p class="site-description motion-element" itemprop="description"></p>
           
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">129</span>
                <span class="site-state-item-name">posts</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">11</span>
                <span class="site-state-item-name">categories</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">100</span>
                <span class="site-state-item-name">tags</span>
              </a>
            </div>
          

        </nav>

        
          <div class="feed-link motion-element">
            <a href="/atom.xml" rel="alternate">
              <i class="fa fa-rss"></i>
              RSS
            </a>
          </div>
        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="https://github.com/tiramisutes" target="_blank" title="GitHub">
                  
                    <i class="fa fa-fw fa-github"></i>
                  
                  GitHub
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="https://twitter.com/xzpgocxx" target="_blank" title="Twitter">
                  
                    <i class="fa fa-fw fa-twitter"></i>
                  
                  Twitter
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://weibo.com/tiramisutesTes" target="_blank" title="Weibo">
                  
                    <i class="fa fa-fw fa-weibo"></i>
                  
                  Weibo
                </a>
              </span>
            
          
        </div>

        
        

        
        
          <div class="links-of-blogroll motion-element links-of-blogroll-inline">
            <div class="links-of-blogroll-title">
              <i class="fa  fa-fw fa-globe"></i>
              Links
            </div>
            <ul class="links-of-blogroll-list">
              
                <li class="links-of-blogroll-item">
                  <a href="https://tiramisutes.gitbooks.io/awk/content/" title="GitBook《awk学习总结》" target="_blank">GitBook《awk学习总结》</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://cotton.hzau.edu.cn/" title="棉花遗传改良团队" target="_blank">棉花遗传改良团队</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.bio-info-trainee.com/" title="生信菜鸟团-深圳大学城" target="_blank">生信菜鸟团-深圳大学城</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.biotrainee.com/forum.php" title="生信技能树" target="_blank">生信技能树</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.iwhgao.com/" title="rabbit gao's blog" target="_blank">rabbit gao's blog</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://huans.github.io/" title="Huans(代谢+GWAS)" target="_blank">Huans(代谢+GWAS)</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.sequencing.uio.no/" title="NSC-sequenceing" target="_blank">NSC-sequenceing</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://blog.jobbole.com/" title="伯乐在线" target="_blank">伯乐在线</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://linux.cn/" title="Linux开源中文社区" target="_blank">Linux开源中文社区</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.r-bloggers.com/" title="R-bloggers" target="_blank">R-bloggers</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://cran.r-project.org/other-docs.html" title="R-resource" target="_blank">R-resource</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html" title="ggplot2" target="_blank">ggplot2</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://lilibei.net/" title="pele' blog" target="_blank">pele' blog</a>
                </li>
              
            </ul>
          </div>
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#sam文件格式"><span class="nav-number">1.</span> <span class="nav-text">sam文件格式</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#比对区解释"><span class="nav-number">1.1.</span> <span class="nav-text">比对区解释</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#sam文件的几个特例解释"><span class="nav-number">2.</span> <span class="nav-text">sam文件的几个特例解释</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Unmapped_reads_统计"><span class="nav-number">2.1.</span> <span class="nav-text">Unmapped reads 统计</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#How_many_different_read_IDs_are_in_the_file?"><span class="nav-number">2.2.</span> <span class="nav-text">How many different read IDs are in the file?</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#How_many_different_read_sequences_are_in_the_file?"><span class="nav-number">2.3.</span> <span class="nav-text">How many different read sequences are in the file?</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#How_many_reads_are_uniquely_mapped?"><span class="nav-number">2.4.</span> <span class="nav-text">How many reads are uniquely mapped?</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#How_many_reads_are_multi-hits?"><span class="nav-number">2.5.</span> <span class="nav-text">How many reads are multi-hits?</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#How_many_alignments_are_reported_for_each_read?"><span class="nav-number">2.6.</span> <span class="nav-text">How many alignments are reported for each read?</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#How_many_different_reference_sequences_are_represented_in_the_file?"><span class="nav-number">2.7.</span> <span class="nav-text">How many different reference sequences are represented in the file?</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#view"><span class="nav-number">3.</span> <span class="nav-text">view</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#sort"><span class="nav-number">4.</span> <span class="nav-text">sort</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#index"><span class="nav-number">5.</span> <span class="nav-text">index</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#faidx"><span class="nav-number">6.</span> <span class="nav-text">faidx</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#flagstat"><span class="nav-number">7.</span> <span class="nav-text">flagstat</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#mpileup"><span class="nav-number">8.</span> <span class="nav-text">mpileup</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#depth"><span class="nav-number">9.</span> <span class="nav-text">depth</span></a></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy;  2014 - 
  <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder"> <a href="/about" target="_blank">tiramisutes</a></span>
</div>

<div class="powered-by">
  Powered by <a class="theme-link" href="https://hexo.io" target="_blank">Hexo</a>
</div>

<div class="theme-info">
  Theme -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next" target="_blank">
    NexT.Muse 
  </a>
<script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js">
                </script>
    <span id="busuanzi_container_site_pv" style='display:none'>
                      &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;<a style="line-height: 1.5;" href="http://c4.gostats.cn/last_guests.xml?id=397056" target="_blank">The total visits <span id="busuanzi_value_site_pv"></span> times</a>
                </span>
    <!-- GoStats JavaScript Based Code -->
                <script type="text/javascript" src="http://gostats.cn/js/counter.js"></script>
                <script type="text/javascript">_gos='c4.gostats.cn';_goa=397056;
                _got=5;_goi=1;_GoStatsRun();</script>
                <noscript><a target="_blank" 
                href="http://gostats.cn"><img src="http://c4.gostats.cn/bin/count/a_397056/t_5/i_1/counter.png" 
                style="border-width:0" /></a></noscript>
                <!-- End GoStats JavaScript Based Code -->  
</div>
<script>
 host3 = window.location.href;
  chinadomai3 = "http://tiramisutes.github.io/";
  if (host3.indexOf(chinadomain3)!=-1)  {

  (function(){
    var bp = document.createElement('script');
    bp.src = '//push.zhanzhang.baidu.com/push.js';
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>
        

        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.1"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.1"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.1"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.1"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.1"></script>



  


  




	

		<script type="text/javascript">
		_hcwp = window._hcwp || [];

		_hcwp.push({widget:"Bloggerstream", widget_id: 91479, selector:".hc-comment-count", label: "{\%COUNT%\}" });

		
		_hcwp.push({widget:"Stream", widget_id: 91479, xid: "2016/03/20/Samtools.html"});
		

		(function() {
		if("HC_LOAD_INIT" in window)return;
		HC_LOAD_INIT = true;
		var lang = (navigator.language || navigator.systemLanguage || navigator.userLanguage || "en").substr(0, 2).toLowerCase();
		var hcc = document.createElement("script"); hcc.type = "text/javascript"; hcc.async = true;
		hcc.src = ("https:" == document.location.protocol ? "https" : "http")+"://w.hypercomments.com/widget/hc/91479/"+lang+"/widget.js";
		var s = document.getElementsByTagName("script")[0];
		s.parentNode.insertBefore(hcc, s.nextSibling);
		})();
		</script>

	












  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (search_path.endsWith("json")) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.1.js"></script>
  <script>AV.initialize("c7x0fPmO8Ci9avDBRuGWHqpG-gzGzoHsz", "iVJ6ST0e1O5xaiIhVwOyhbf3");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  
<script>
(function(){
    var bp = document.createElement('script');
    var curProtocol = window.location.protocol.split(':')[0];
    if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
    }
    else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
    }
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


  
  


  
  <script type="text/javascript" src="/js/src/js.cookie.js?v=5.1.1"></script>
  <script type="text/javascript" src="/js/src/scroll-cookie.js?v=5.1.1"></script>


  

</body>
</html>
