<!doctype html>



  


<html class="theme-next muse use-motion" lang="default">
<head>
  <meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>



  
  
  <link rel="stylesheet" media="all" href="/lib/Han/dist/han.min.css?v=3.3">




<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="google-site-verification" content="b12Zu40L63PLgCtcn0IW8KCNRolmyQlaF22tf4rRLEg" />













  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css" />




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="http://fonts.useso.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css" />

<link href="/css/main.css?v=5.1.1" rel="stylesheet" type="text/css" />


  <meta name="keywords" content="FASTX," />





  <link rel="alternate" href="/atom.xml" title="hope" type="application/atom+xml" />




  <link rel="shortcut icon" type="image/x-icon" href="/images/favicon.ico?v=5.1.1" />






<meta name="description" content="Here you’ll find a short description and examples of how to use the FASTX-toolkit from the command line.
Command Line ArgumentsMost tools show usage information with -h.Tools can read from STDIN and w">
<meta property="og:type" content="article">
<meta property="og:title" content="FASTX-Toolkit">
<meta property="og:url" content="http://tiramisutes.github.io/2015/07/24/FASTX-Toolkit.html">
<meta property="og:site_name" content="hope">
<meta property="og:description" content="Here you’ll find a short description and examples of how to use the FASTX-toolkit from the command line.
Command Line ArgumentsMost tools show usage information with -h.Tools can read from STDIN and w">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/bc54_quality.png">
<meta property="og:image" content="http://7xk19o.com1.z0.glb.clouddn.com/bc54_nuc.png">
<meta property="og:updated_time" content="2015-07-25T12:44:45.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="FASTX-Toolkit">
<meta name="twitter:description" content="Here you’ll find a short description and examples of how to use the FASTX-toolkit from the command line.
Command Line ArgumentsMost tools show usage information with -h.Tools can read from STDIN and w">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":0,"b2t":false,"scrollpercent":false},
    fancybox: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://tiramisutes.github.io/2015/07/24/FASTX-Toolkit.html"/>





  <title>FASTX-Toolkit | hope</title>
</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  





  <script type="text/javascript">
    var _hmt = _hmt || [];
    (function() {
      var hm = document.createElement("script");
      hm.src = "https://hm.baidu.com/hm.js?08b17c69c7ba460d9ab5f188a048ee6d";
      var s = document.getElementsByTagName("script")[0];
      s.parentNode.insertBefore(hm, s);
    })();
  </script>











  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/"  class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">hope</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <h1 class="site-subtitle" itemprop="description">A new day is coming,whether we like it or not. The question is will you control it,or will it control you?</h1>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br />
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br />
            
            Archives
          </a>
        </li>
      
        
        <li class="menu-item menu-item-code">
          <a href="/code" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-code"></i> <br />
            
            Code
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br />
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br />
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-footprint">
          <a href="/footprint" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-image"></i> <br />
            
            Footprint
          </a>
        </li>
      
        
        <li class="menu-item menu-item-about">
          <a href="/about" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-user"></i> <br />
            
            About
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br />
            
            Search
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off"
             placeholder="Searching..." spellcheck="false"
             type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal " itemscope itemtype="http://schema.org/Article">
    <link itemprop="mainEntityOfPage" href="http://tiramisutes.github.io/2015/07/24/FASTX-Toolkit.html">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="tiramisutes">
      <meta itemprop="description" content="">
      <meta itemprop="image" content="/images/author.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="hope">
    </span>

    
      <header class="post-header">

        
        
          <h2 class="post-title" itemprop="name headline">FASTX-Toolkit</h2>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2015-07-24T22:51:33+08:00">
                2015-07-24
              </time>
            

            

            
          </span>

          
            <span class="post-category" >
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/Bioinformatics/" itemprop="url" rel="index">
                    <span itemprop="name">Bioinformatics</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
            <!--noindex-->
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2015/07/24/FASTX-Toolkit.html#comments" itemprop="discussionUrl">
                  <span class="post-comments-count hc-comment-count" data-xid="2015/07/24/FASTX-Toolkit.html" itemprop="commentsCount"></span>
                </a>
              </span>
              <!--/noindex-->
            
          

          
          
             <span id="/2015/07/24/FASTX-Toolkit.html" class="leancloud_visitors" data-flag-title="FASTX-Toolkit">
               <span class="post-meta-divider">|</span>
               <span class="post-meta-item-icon">
                 <i class="fa fa-eye"></i>
               </span>
               
                 <span class="post-meta-item-text">Visitors </span>
               
                 <span class="leancloud-visitors-count"></span>
             </span>
          

          

          
            <div class="post-wordcount">
              
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Words count in article</span>
                
                <span title="Words count in article">
                  5,421
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">Reading time</span>
                
                <span title="Reading time">
                  34
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    <div class="post-body han-init-context" itemprop="articleBody">

      
      

      
        <p>Here you’ll find a short description and examples of how to use the FASTX-toolkit from the command line.</p>
<h2 id="Command_Line_Arguments">Command Line Arguments</h2><p>Most tools show usage information with -h.<br>Tools can read from STDIN and write to STDOUT, or<br>from a specific input file (-i) and specific output file (-o).<br>Tools can operate silently (producing no output if everything was OK), or<br>print a short summary (-v).<br>If output goes to STDOUT, the summary will be printed to STDERR.<br>If output goes to a file, the summary will be printed to STDOUT.<br>Some tools can compress the output with GZIP (-z).</p>
<h3 id="FASTQ-to-FASTA">FASTQ-to-FASTA</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">$ fastq_to_fasta -h</span><br><span class="line">usage: fastq_to_fasta [-h] [-r] [-n] [-v] [-z] [-i INFILE] [-o OUTFILE]</span><br><span class="line"></span><br><span class="line">version <span class="number">0.0</span>.<span class="number">6</span></span><br><span class="line">   [-h]         = This helpful <span class="built_in">help</span> screen.</span><br><span class="line">   [-r]         = Rename sequence identifiers to numbers.</span><br><span class="line">   [-n]         = keep sequences with unknown (N) nucleotides.</span><br><span class="line">		  Default is to discard such sequences.</span><br><span class="line">   [-v]         = Verbose - report number of sequences.</span><br><span class="line">		  If [-o] is specified,  report will be printed to STDOUT.</span><br><span class="line">		  If [-o] is not specified (and output goes to STDOUT),</span><br><span class="line">		  report will be printed to STDERR.</span><br><span class="line">   [-z]         = Compress output with GZIP.</span><br><span class="line">   [-i INFILE]  = FASTA/Q input file. default is STDIN.</span><br><span class="line">   [-o OUTFILE] = FASTA output file. default is STDOUT.</span><br></pre></td></tr></table></figure>
<h3 id="FASTX_Statistics">FASTX Statistics</h3><pre><code>$ fastx_quality_stats -<span class="literal">h</span>
usage: fastx_quality_stats [-<span class="keyword">h</span>] [-i <span class="keyword">INFILE</span>] [-o <span class="keyword">OUTFILE</span>]

<span class="keyword">version</span> 0.0.6 (C) 2008 <span class="keyword">by</span> Assaf Gordon (gordon@cshl.edu)
   [-<span class="keyword">h</span>] = This helpful <span class="keyword">help</span> screen.
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> <span class="keyword">file</span>. default is STDIN.
                  <span class="keyword">If</span> FASTA <span class="keyword">file</span> is given, only nucleotides
          distribution is calculated (there's <span class="keyword">no</span> quality info).
   [-o <span class="keyword">OUTFILE</span>] = TEXT output <span class="keyword">file</span>. default is STDOUT.

The output TEXT <span class="keyword">file</span> will have the following fields (<span class="keyword">one</span> row per column):
    column    = column number (1 to 36 <span class="keyword">for</span> a 36-cycles <span class="keyword">read</span> solexa <span class="keyword">file</span>)
    <span class="keyword">count</span>   = number of bases found <span class="keyword">in</span> this column.
    min     = Lowest quality <span class="keyword">score</span> value found <span class="keyword">in</span> this column.
    max     = Highest quality <span class="keyword">score</span> value found <span class="keyword">in</span> this column.
    <span class="keyword">sum</span>     = <span class="keyword">Sum</span> of quality <span class="keyword">score</span> values <span class="keyword">for</span> this column.
    <span class="keyword">mean</span>    = <span class="keyword">Mean</span> quality <span class="keyword">score</span> value <span class="keyword">for</span> this column.
    Q1    = 1st quartile quality <span class="keyword">score</span>.
    med    = <span class="keyword">Median</span> quality <span class="keyword">score</span>.
    Q3    = 3rd quartile quality <span class="keyword">score</span>.
    IQR    = Inter-Quartile <span class="keyword">range</span> (Q3-Q1).
    lW    = 'Left-Whisker' value (<span class="keyword">for</span> boxplotting).
    rW    = 'Right-Whisker' value (<span class="keyword">for</span> boxplotting).
    A_Count    = <span class="keyword">Count</span> of 'A' nucleotides found <span class="keyword">in</span> this column.
    C_Count    = <span class="keyword">Count</span> of 'C' nucleotides found <span class="keyword">in</span> this column.
    G_Count    = <span class="keyword">Count</span> of '<span class="keyword">G</span>' nucleotides found <span class="keyword">in</span> this column.
    T_Count    = <span class="keyword">Count</span> of 'T' nucleotides found <span class="keyword">in</span> this column.
    N_Count = <span class="keyword">Count</span> of '<span class="keyword">N</span>' nucleotides found <span class="keyword">in</span> this column.
    max-<span class="keyword">count</span> = max. number of bases (<span class="keyword">in</span> all cycles)
</code></pre><h3 id="FASTQ_Quality_Chart">FASTQ Quality Chart</h3><pre><code>$ fastq_quality_boxplot_graph.<span class="keyword">sh</span> -<span class="literal">h</span>
Solexa-Quality BoxPlot plotter
Generates a solexa quality <span class="keyword">score</span> box-<span class="keyword">plot</span> <span class="keyword">graph</span> 

Usage: /usr/<span class="keyword">local</span>/bin/fastq_quality_boxplot_graph.<span class="keyword">sh</span> [-i <span class="keyword">INPUT</span>.TXT] [-t TITLE] [-p] [-o OUTPUT]

  [-p]           - <span class="keyword">Generate</span> PostScript (.PS) <span class="keyword">file</span>. Default is PNG image.
  [-i <span class="keyword">INPUT</span>.TXT] - <span class="keyword">Input</span> <span class="keyword">file</span>. Should be the output of <span class="string">"solexa_quality_statistics"</span> <span class="keyword">program</span>.
  [-o OUTPUT]    - Output <span class="keyword">file</span> name. default is STDOUT.
  [-t TITLE]     - Title (usually the solexa <span class="keyword">file</span> name) - will be plotted <span class="keyword">on</span> the <span class="keyword">graph</span>.
</code></pre><h3 id="FASTA/Q_Nucleotide_Distribution">FASTA/Q Nucleotide Distribution</h3><pre><code>$ fastx_nucleotide_distribution_graph.<span class="keyword">sh</span> -<span class="literal">h</span>
FASTA/Q Nucleotide Distribution Plotter

Usage: /usr/<span class="keyword">local</span>/bin/fastx_nucleotide_distribution_graph.<span class="keyword">sh</span> [-i <span class="keyword">INPUT</span>.TXT] [-t TITLE] [-p] [-o OUTPUT]

  [-p]           - <span class="keyword">Generate</span> PostScript (.PS) <span class="keyword">file</span>. Default is PNG image.
  [-i <span class="keyword">INPUT</span>.TXT] - <span class="keyword">Input</span> <span class="keyword">file</span>. Should be the output of <span class="string">"fastx_quality_statistics"</span> <span class="keyword">program</span>.
  [-o OUTPUT]    - Output <span class="keyword">file</span> name. default is STDOUT.
  [-t TITLE]     - Title - will be plotted <span class="keyword">on</span> the <span class="keyword">graph</span>.
</code></pre><h3 id="FASTA/Q_Clipper">FASTA/Q Clipper</h3><pre><code>$ fastx_clipper -h
usage: fastx_clipper [-h] [-a ADAPTER] [-D] [-l N] [-n] [-d N] [-c] [-C] [-o] [-v] [-z] [-i INFILE] [-o OUTFILE]

version 0.0.6
   [-h]         = This helpful <span class="operator"><span class="keyword">help</span> screen.
   [-a ADAPTER] = ADAPTER <span class="keyword">string</span>. <span class="keyword">default</span> <span class="keyword">is</span> CCTTAAGG (dummy adapter).
   [-l N]       = discard sequences shorter than N nucleotides. <span class="keyword">default</span> <span class="keyword">is</span> <span class="number">5.</span>
   [-d N]       = Keep the adapter <span class="keyword">and</span> N bases <span class="keyword">after</span> it.
          (<span class="keyword">using</span> <span class="string">'-d 0'</span> <span class="keyword">is</span> the same <span class="keyword">as</span> <span class="keyword">not</span> <span class="keyword">using</span> <span class="string">'-d'</span> <span class="keyword">at</span> <span class="keyword">all</span>. which <span class="keyword">is</span> the <span class="keyword">default</span>).
   [-c]         = Discard non-clipped sequences (i.e. - keep <span class="keyword">only</span> sequences which contained the adapter).
   [-C]         = Discard clipped sequences (i.e. - keep <span class="keyword">only</span> sequences which did <span class="keyword">not</span> contained the adapter).
   [-k]         = Report Adapter-<span class="keyword">Only</span> sequences.
   [-n]         = keep sequences <span class="keyword">with</span> <span class="keyword">unknown</span> (N) nucleotides. <span class="keyword">default</span> <span class="keyword">is</span> <span class="keyword">to</span> discard such sequences.
   [-v]         = Verbose - report <span class="built_in">number</span> <span class="keyword">of</span> sequences.
          <span class="keyword">If</span> [-o] <span class="keyword">is</span> specified,  report will be printed <span class="keyword">to</span> STDOUT.
          <span class="keyword">If</span> [-o] <span class="keyword">is</span> <span class="keyword">not</span> specified (<span class="keyword">and</span> <span class="keyword">output</span> goes <span class="keyword">to</span> STDOUT),
          report will be printed <span class="keyword">to</span> STDERR.
   [-z]         = <span class="keyword">Compress</span> <span class="keyword">output</span> <span class="keyword">with</span> GZIP.
   [-D]        = DEBUG <span class="keyword">output</span>.
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDIN.
   [-o <span class="keyword">OUTFILE</span>] = FASTA/Q <span class="keyword">output</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDOUT.</span>
</code></pre><h3 id="FASTA/Q_Renamer">FASTA/Q Renamer</h3><pre><code>$ fastx_renamer -<span class="literal">h</span>
usage: fastx_renamer [-<span class="keyword">n</span> <span class="keyword">TYPE</span>] [-<span class="keyword">h</span>] [-z] [-v] [-i <span class="keyword">INFILE</span>] [-o <span class="keyword">OUTFILE</span>]
Part of FASTX Toolkit 0.0.10 <span class="keyword">by</span> A. Gordon (gordon@cshl.edu)

   [-<span class="keyword">n</span> <span class="keyword">TYPE</span>]    = <span class="keyword">rename</span> <span class="keyword">type</span>:
          SEQ - <span class="keyword">use</span> the nucleotides sequence <span class="keyword">as</span> the name.
          <span class="keyword">COUNT</span> - <span class="keyword">use</span> simply counter <span class="keyword">as</span> the name.
   [-<span class="keyword">h</span>]         = This helpful <span class="keyword">help</span> screen.
   [-z]         = <span class="keyword">Compress</span> output with GZIP.
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> <span class="keyword">file</span>. default is STDIN.
   [-o <span class="keyword">OUTFILE</span>] = FASTA/Q output <span class="keyword">file</span>. default is STDOUT.
</code></pre><h3 id="FASTA/Q_Trimmer">FASTA/Q Trimmer</h3><pre><code>$ fastx_trimmer -h
usage: fastx_trimmer [-h] [-f N] [-l N] [-z] [-v] [-i INFILE] [-o OUTFILE]

version 0.0.6
   [-h]         = This helpful <span class="operator"><span class="keyword">help</span> screen.
   [-f N]       = <span class="keyword">First</span> base <span class="keyword">to</span> keep. <span class="keyword">Default</span> <span class="keyword">is</span> <span class="number">1</span> (=<span class="keyword">first</span> base).
   [-l N]       = <span class="keyword">Last</span> base <span class="keyword">to</span> keep. <span class="keyword">Default</span> <span class="keyword">is</span> entire <span class="keyword">read</span>.
   [-z]         = <span class="keyword">Compress</span> <span class="keyword">output</span> <span class="keyword">with</span> GZIP.
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDIN.
   [-o <span class="keyword">OUTFILE</span>] = FASTA/Q <span class="keyword">output</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDOUT.</span>
</code></pre><h3 id="FASTA/Q_Collapser">FASTA/Q Collapser</h3><pre><code>$ fastx_collapser -<span class="literal">h</span>
usage: fastx_collapser [-<span class="keyword">h</span>] [-v] [-i <span class="keyword">INFILE</span>] [-o <span class="keyword">OUTFILE</span>]

<span class="keyword">version</span> 0.0.6
   [-<span class="keyword">h</span>]         = This helpful <span class="keyword">help</span> screen.
   [-v]         = verbose: <span class="keyword">print</span> short summary of <span class="keyword">input</span>/output counts
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> <span class="keyword">file</span>. default is STDIN.
   [-o <span class="keyword">OUTFILE</span>] = FASTA/Q output <span class="keyword">file</span>. default is STDOUT.
</code></pre><h3 id="FASTQ/A_Artifacts_Filter">FASTQ/A Artifacts Filter</h3><pre><code>$ fastx_artifacts_filter -h
usage: fastq_artifacts_filter [-h] [-v] [-z] [-i INFILE] [-o OUTFILE]

version 0.0.6
   [-h]         = This helpful <span class="operator"><span class="keyword">help</span> screen.
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDIN.
   [-o <span class="keyword">OUTFILE</span>] = FASTA/Q <span class="keyword">output</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDOUT.
   [-z]         = <span class="keyword">Compress</span> <span class="keyword">output</span> <span class="keyword">with</span> GZIP.
   [-v]         = Verbose - report <span class="built_in">number</span> <span class="keyword">of</span> processed reads.
          <span class="keyword">If</span> [-o] <span class="keyword">is</span> specified,  report will be printed <span class="keyword">to</span> STDOUT.
          <span class="keyword">If</span> [-o] <span class="keyword">is</span> <span class="keyword">not</span> specified (<span class="keyword">and</span> <span class="keyword">output</span> goes <span class="keyword">to</span> STDOUT),
          report will be printed <span class="keyword">to</span> STDERR.</span>
</code></pre><h3 id="FASTQ_Quality_Filter">FASTQ Quality Filter</h3><pre><code>$ <span class="atom">fastq_quality_filter</span> -<span class="atom">h</span>
<span class="atom">usage</span>: <span class="atom">fastq_quality_filter</span> [-<span class="atom">h</span>] [-<span class="atom">v</span>] [-<span class="atom">q</span> <span class="name">N</span>] [-<span class="atom">p</span> <span class="name">N</span>] [-<span class="atom">z</span>] [-<span class="atom">i</span> <span class="name">INFILE</span>] [-<span class="atom">o</span> <span class="name">OUTFILE</span>]

<span class="atom">version</span> <span class="number">0.0</span>.6
   [-<span class="atom">h</span>]         = <span class="name">This</span> <span class="atom">helpful</span> <span class="atom">help</span> <span class="atom">screen</span>.
   [-<span class="atom">q</span> <span class="name">N</span>]       = <span class="name">Minimum</span> <span class="atom">quality</span> <span class="atom">score</span> <span class="atom">to</span> <span class="atom">keep</span>.
   [-<span class="atom">p</span> <span class="name">N</span>]       = <span class="name">Minimum</span> <span class="atom">percent</span> <span class="atom">of</span> <span class="atom">bases</span> <span class="atom">that</span> <span class="atom">must</span> <span class="atom">have</span> [-<span class="atom">q</span>] <span class="atom">quality</span>.
   [-<span class="atom">z</span>]         = <span class="name">Compress</span> <span class="atom">output</span> <span class="atom">with</span> <span class="name">GZIP</span>.
   [-<span class="atom">i</span> <span class="name">INFILE</span>]  = <span class="name">FASTA</span>/<span class="name">Q</span> <span class="atom">input</span> <span class="atom">file</span>. <span class="atom">default</span> <span class="atom">is</span> <span class="name">STDIN</span>.
   [-<span class="atom">o</span> <span class="name">OUTFILE</span>] = <span class="name">FASTA</span>/<span class="name">Q</span> <span class="atom">output</span> <span class="atom">file</span>. <span class="atom">default</span> <span class="atom">is</span> <span class="name">STDOUT</span>.
   [-<span class="atom">v</span>]         = <span class="name">Verbose</span> - <span class="atom">report</span> <span class="atom">number</span> <span class="atom">of</span> <span class="atom">sequences</span>.
          <span class="name">If</span> [-<span class="atom">o</span>] <span class="atom">is</span> <span class="atom">specified</span>,  <span class="atom">report</span> <span class="atom">will</span> <span class="atom">be</span> <span class="atom">printed</span> <span class="atom">to</span> <span class="name">STDOUT</span>.
          <span class="name">If</span> [-<span class="atom">o</span>] <span class="atom">is</span> <span class="atom">not</span> <span class="atom">specified</span> (<span class="atom">and</span> <span class="atom">output</span> <span class="atom">goes</span> <span class="atom">to</span> <span class="name">STDOUT</span>),
          <span class="atom">report</span> <span class="atom">will</span> <span class="atom">be</span> <span class="atom">printed</span> <span class="atom">to</span> <span class="name">STDERR</span>.
</code></pre><h3 id="FASTQ/A_Reverse_Complement">FASTQ/A Reverse Complement</h3><pre><code>$ <span class="atom">fastx_reverse_complement</span> -<span class="atom">h</span>
<span class="atom">usage</span>: <span class="atom">fastx_reverse_complement</span> [-<span class="atom">h</span>] [-<span class="atom">r</span>] [-<span class="atom">z</span>] [-<span class="atom">v</span>] [-<span class="atom">i</span> <span class="name">INFILE</span>] [-<span class="atom">o</span> <span class="name">OUTFILE</span>]

<span class="atom">version</span> <span class="number">0.0</span>.6
   [-<span class="atom">h</span>]         = <span class="name">This</span> <span class="atom">helpful</span> <span class="atom">help</span> <span class="atom">screen</span>.
   [-<span class="atom">z</span>]         = <span class="name">Compress</span> <span class="atom">output</span> <span class="atom">with</span> <span class="name">GZIP</span>.
   [-<span class="atom">i</span> <span class="name">INFILE</span>]  = <span class="name">FASTA</span>/<span class="name">Q</span> <span class="atom">input</span> <span class="atom">file</span>. <span class="atom">default</span> <span class="atom">is</span> <span class="name">STDIN</span>.
   [-<span class="atom">o</span> <span class="name">OUTFILE</span>] = <span class="name">FASTA</span>/<span class="name">Q</span> <span class="atom">output</span> <span class="atom">file</span>. <span class="atom">default</span> <span class="atom">is</span> <span class="name">STDOUT</span>.
</code></pre><h3 id="FASTA_Formatter">FASTA Formatter</h3><pre><code>$ fasta_formatter -h
usage: fasta_formatter [-h] [-i INFILE] [-o OUTFILE] [-w N] [-t] [-e]
Part of FASTX Toolkit 0.0.7 by gordon@cshl.edu

   [-h]         = This helpful <span class="operator"><span class="keyword">help</span> screen.
   [-i <span class="keyword">INFILE</span>]  = FASTA/Q <span class="keyword">input</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDIN.
   [-o <span class="keyword">OUTFILE</span>] = FASTA/Q <span class="keyword">output</span> file. <span class="keyword">default</span> <span class="keyword">is</span> STDOUT.
   [-w N]       = <span class="keyword">max</span>. sequence line width <span class="keyword">for</span> <span class="keyword">output</span> FASTA file.
          <span class="keyword">When</span> ZERO (the <span class="keyword">default</span>), sequence <span class="keyword">lines</span> will <span class="keyword">NOT</span> be wrapped -
          <span class="keyword">all</span> nucleotides <span class="keyword">of</span> <span class="keyword">each</span> sequences will appear <span class="keyword">on</span> a single 
          line (good <span class="keyword">for</span> scripting).
   [-t]         = <span class="keyword">Output</span> tabulated <span class="keyword">format</span> (instead <span class="keyword">of</span> FASTA <span class="keyword">format</span>).
          Sequence-Identifiers will be <span class="keyword">on</span> <span class="keyword">first</span> <span class="keyword">column</span>,
          Nucleotides will appear <span class="keyword">on</span> <span class="keyword">second</span> <span class="keyword">column</span> (<span class="keyword">as</span> single line).
   [-e]         = <span class="keyword">Output</span> empty sequences (<span class="keyword">default</span> <span class="keyword">is</span> <span class="keyword">to</span> discard them).
          Empty sequences <span class="keyword">are</span> ones who have <span class="keyword">only</span> a sequence identifier,
          but <span class="keyword">not</span> actual nucleotides.

<span class="keyword">Input</span> Example:
   &gt;MY-ID
   AAAAAGGGGG
   CCCCCTTTTT
   AGCTN

<span class="keyword">Output</span> example <span class="keyword">with</span> unlimited line width [-w <span class="number">0</span>]:
   &gt;MY-ID
   AAAAAGGGGGCCCCCTTTTTAGCTN

<span class="keyword">Output</span> example <span class="keyword">with</span> <span class="keyword">max</span>. line width=<span class="number">7</span> [-w <span class="number">7</span>]:
   &gt;MY-ID
   AAAAAGG
   GGGTTTT
   TCCCCCA
   GCTN

<span class="keyword">Output</span> example <span class="keyword">with</span> tabular <span class="keyword">output</span> [-t]:
   MY-ID    AAAAAGGGGGCCCCCTTTTAGCTN

example <span class="keyword">of</span> empty sequence:
(will be discarded unless [-e] <span class="keyword">is</span> used)
  &gt;REGULAR-SEQUENCE-<span class="number">1</span>
  AAAGGGTTTCCC
  &gt;EMPTY-SEQUENCE
  &gt;REGULAR-SEQUENCE-<span class="number">2</span>
  AAGTAGTAGTAGTAGT
  GTATTTTATAT</span>
</code></pre><h3 id="FASTA_Nucleotides_Changer">FASTA Nucleotides Changer</h3><pre><code>$ <span class="atom">fasta_nucleotide_changer</span> -<span class="atom">h</span>
<span class="atom">usage</span>: <span class="atom">fasta_nucleotide_changer</span> [-<span class="atom">h</span>] [-<span class="atom">z</span>] [-<span class="atom">v</span>] [-<span class="atom">i</span> <span class="name">INFILE</span>] [-<span class="atom">o</span> <span class="name">OUTFILE</span>] [-<span class="atom">r</span>] [-<span class="atom">d</span>]

<span class="atom">version</span> <span class="number">0.0</span>.7
   [-<span class="atom">h</span>]         = <span class="name">This</span> <span class="atom">helpful</span> <span class="atom">help</span> <span class="atom">screen</span>.
   [-<span class="atom">z</span>]         = <span class="name">Compress</span> <span class="atom">output</span> <span class="atom">with</span> <span class="name">GZIP</span>.
   [-<span class="atom">v</span>]         = <span class="name">Verbose</span> <span class="atom">mode</span>. <span class="name">Prints</span> <span class="atom">a</span> <span class="atom">short</span> <span class="atom">summary</span>.
          <span class="atom">with</span> [-<span class="atom">o</span>], <span class="atom">summary</span> <span class="atom">is</span> <span class="atom">printed</span> <span class="atom">to</span> <span class="name">STDOUT</span>.
          <span class="name">Otherwise</span>, <span class="atom">summary</span> <span class="atom">is</span> <span class="atom">printed</span> <span class="atom">to</span> <span class="name">STDERR</span>.
   [-<span class="atom">i</span> <span class="name">INFILE</span>]  = <span class="name">FASTA</span>/<span class="name">Q</span> <span class="atom">input</span> <span class="atom">file</span>. <span class="atom">default</span> <span class="atom">is</span> <span class="name">STDIN</span>.
   [-<span class="atom">o</span> <span class="name">OUTFILE</span>] = <span class="name">FASTA</span>/<span class="name">Q</span> <span class="atom">output</span> <span class="atom">file</span>. <span class="atom">default</span> <span class="atom">is</span> <span class="name">STDOUT</span>.
   [-<span class="atom">r</span>]         = <span class="name">DNA</span>-<span class="atom">to</span>-<span class="name">RNA</span> <span class="atom">mode</span> - <span class="atom">change</span> <span class="name">T</span><span class="string">'s into U'</span><span class="atom">s</span>.
   [-<span class="atom">d</span>]         = <span class="name">RNA</span>-<span class="atom">to</span>-<span class="name">DNA</span> <span class="atom">mode</span> - <span class="atom">change</span> <span class="name">U</span><span class="string">'s into T'</span><span class="atom">s</span>.
</code></pre><h3 id="FASTA_Clipping_Histogram">FASTA Clipping Histogram</h3><pre><code>$ fasta_clipping_histogram<span class="class">.pl</span>
Create <span class="tag">a</span> Linker Clipping Information Histogram

usage: fasta_clipping_histogram<span class="class">.pl</span> INPUT_FILE<span class="class">.FA</span> OUTPUT_FILE<span class="class">.PNG</span>

    INPUT_FILE<span class="class">.FA</span>   = <span class="tag">input</span> file (<span class="keyword">in</span> FASTA format, can be GZIPped)
    OUTPUT_FILE<span class="class">.PNG</span> = histogram image
</code></pre><h3 id="FASTX_Barcode_Splitter">FASTX Barcode Splitter</h3><pre><code>$ fastx_barcode_splitter.pl
Barcode Splitter, <span class="keyword">by</span> Assaf Gordon (gordon@cshl.edu), <span class="number">11</span>sep2008

This program reads FASTA/FASTQ <span class="type">file</span> <span class="keyword">and</span> splits <span class="keyword">it</span> <span class="keyword">into</span> several smaller files,
Based <span class="function_start"><span class="keyword">on</span></span> barcode matching.
FASTA/FASTQ data <span class="keyword">is</span> <span class="command">read</span> <span class="keyword">from</span> STDIN (format <span class="keyword">is</span> auto-detected.)
Output files will be writen <span class="keyword">to</span> disk.
Summary will be printed <span class="keyword">to</span> STDOUT.

usage: /usr/<span class="keyword">local</span>/bin/fastx_barcode_splitter.pl <span class="comment">--bcfile FILE --prefix PREFIX [--suffix SUFFIX] [--bol|--eol] </span>
     [<span class="comment">--mismatches N] [--exact] [--partial N] [--help] [--quiet] [--debug]</span>

Arguments:

<span class="comment">--bcfile FILE    - Barcodes file name. (see explanation below.)</span>
<span class="comment">--prefix PREFIX    - File prefix. will be added to the output files. Can be used</span>
          <span class="keyword">to</span> specify output directories.
<span class="comment">--suffix SUFFIX    - File suffix (optional). Can be used to specify file</span>
          extensions.
<span class="comment">--bol        - Try to match barcodes at the BEGINNING of sequences.</span>
          (What biologists would call <span class="keyword">the</span> <span class="number">5</span>' <span class="keyword">end</span>, <span class="keyword">and</span> programmers
          would call index <span class="number">0.</span>)
<span class="comment">--eol        - Try to match barcodes at the END of sequences.</span>
          (What biologists would call <span class="keyword">the</span> <span class="number">3</span>' <span class="keyword">end</span>, <span class="keyword">and</span> programmers
          would call <span class="keyword">the</span> <span class="keyword">end</span> <span class="keyword">of</span> <span class="keyword">the</span> <span class="type">string</span>.)
          NOTE: one <span class="keyword">of</span> <span class="comment">--bol, --eol must be specified, but not both.</span>
<span class="comment">--mismatches N    - Max. number of mismatches allowed. default is 1.</span>
<span class="comment">--exact        - Same as '--mismatches 0'. If both --exact and --mismatches </span>
          are specified, '<span class="comment">--exact' takes precedence.</span>
<span class="comment">--partial N    - Allow partial overlap of barcodes. (see explanation below.)</span>
          (Default <span class="keyword">is</span> <span class="keyword">not</span> partial matching)
<span class="comment">--quiet        - Don't print counts and summary at the end of the run.</span>
          (Default <span class="keyword">is</span> <span class="keyword">to</span> print.)
<span class="comment">--debug        - Print lots of useless debug information to STDERR.</span>
<span class="comment">--help        - This helpful help screen.</span>

Example (Assuming 's_2_100.txt' <span class="keyword">is</span> a FASTQ <span class="type">file</span>, 'mybarcodes.txt' <span class="keyword">is</span> 

<span class="keyword">the</span> barcodes <span class="type">file</span>):

   $ cat s_2_100.txt | /usr/<span class="keyword">local</span>/bin/fastx_barcode_splitter.pl <span class="comment">--bcfile mybarcodes.txt --bol --mismatches 2 \</span>
    <span class="comment">--prefix /tmp/bla_ --suffix ".txt"</span>

Barcode <span class="type">file</span> format
<span class="comment">-------------------</span>
Barcode files are simple <span class="type">text</span> files. Each line should <span class="keyword">contain</span> an identifier 
(descriptive <span class="property">name</span> <span class="keyword">for</span> <span class="keyword">the</span> barcode), <span class="keyword">and</span> <span class="keyword">the</span> barcode itself (A/C/G/T), 
separated <span class="keyword">by</span> a TAB <span class="property">character</span>. Example:

    <span class="comment">#This line is a comment (starts with a 'number' sign)</span>
    BC1 GATCT
    BC2 ATCGT
    BC3 GTGAT
    BC4 TGTCT

For each barcode, a new FASTQ <span class="type">file</span> will be created (<span class="keyword">with</span> <span class="keyword">the</span> barcode's 
identifier <span class="keyword">as</span> part <span class="keyword">of</span> <span class="keyword">the</span> <span class="type">file</span> <span class="property">name</span>). Sequences matching <span class="keyword">the</span> barcode 
will be stored <span class="keyword">in</span> <span class="keyword">the</span> appropriate <span class="type">file</span>.

Running <span class="keyword">the</span> <span class="keyword">above</span> example (assuming <span class="string">"mybarcodes.txt"</span> <span class="keyword">contains</span> <span class="keyword">the</span> <span class="keyword">above</span> 
barcodes), will create <span class="keyword">the</span> following files:
    /tmp/bla_BC1.txt
    /tmp/bla_BC2.txt
    /tmp/bla_BC3.txt
    /tmp/bla_BC4.txt
    /tmp/bla_unmatched.txt
The 'unmatched' <span class="type">file</span> will <span class="keyword">contain</span> all sequences <span class="keyword">that</span> didn't match any barcode.

Barcode matching
<span class="comment">----------------</span>

** Without partial matching:

Count mismatches <span class="keyword">between</span> <span class="keyword">the</span> FASTA/Q sequences <span class="keyword">and</span> <span class="keyword">the</span> barcodes.
The barcode which matched <span class="keyword">with</span> <span class="keyword">the</span> lowest mismatches <span class="command">count</span> (providing <span class="keyword">the</span>
<span class="command">count</span> <span class="keyword">is</span> small <span class="keyword">or</span> <span class="keyword">equal</span> <span class="keyword">to</span> '<span class="comment">--mismatches N') 'gets' the sequences.</span>

Example (using <span class="keyword">the</span> <span class="keyword">above</span> barcodes):
Input Sequence:
    GATTTACTATGTAAAGATAGAAGGAATAAGGTGAAG

Matching <span class="keyword">with</span> '<span class="comment">--bol --mismatches 1':</span>
   GATTTACTATGTAAAGATAGAAGGAATAAGGTGAAG
   GATCT (<span class="number">1</span> mismatch, BC1)
   ATCGT (<span class="number">4</span> mismatches, BC2)
   GTGAT (<span class="number">3</span> mismatches, BC3)
   TGTCT (<span class="number">3</span> mismatches, BC4)

This sequence will be classified <span class="keyword">as</span> 'BC1' (<span class="keyword">it</span> has <span class="keyword">the</span> lowest mismatch <span class="command">count</span>).
If '<span class="comment">--exact' or '--mismatches 0' were specified, this sequence would be </span>
classified <span class="keyword">as</span> 'unmatched' (because, although BC1 had <span class="keyword">the</span> lowest mismatch <span class="command">count</span>,
<span class="keyword">it</span> <span class="keyword">is</span> <span class="keyword">above</span> <span class="keyword">the</span> maximum allowed mismatches).

Matching <span class="keyword">with</span> '<span class="comment">--eol' (end of line) does the same, but from the other side</span>
<span class="keyword">of</span> <span class="keyword">the</span> sequence.

** With partial matching (very similar <span class="keyword">to</span> indels):

Same <span class="keyword">as</span> <span class="keyword">above</span>, <span class="keyword">with</span> <span class="keyword">the</span> following addition: barcodes are also checked <span class="keyword">for</span>
partial overlap (<span class="type">number</span> <span class="keyword">of</span> allowed non-overlapping bases <span class="keyword">is</span> '<span class="comment">--partial N').</span>

Example:
Input sequence <span class="keyword">is</span> ATTTACTATGTAAAGATAGAAGGAATAAGGTGAAG
(Same <span class="keyword">as</span> <span class="keyword">above</span>, <span class="keyword">but</span> note <span class="keyword">the</span> missing 'G' <span class="keyword">at</span> <span class="keyword">the</span> <span class="keyword">beginning</span>.)

Matching (<span class="keyword">without</span> partial overlapping) <span class="keyword">against</span> BC1 yields <span class="number">4</span> mismatches:
   ATTTACTATGTAAAGATAGAAGGAATAAGGTGAAG
   GATCT (<span class="number">4</span> mismatches)

Partial overlapping would also <span class="keyword">try</span> <span class="keyword">the</span> following match:
   -ATTTACTATGTAAAGATAGAAGGAATAAGGTGAAG
   GATCT (<span class="number">1</span> mismatch)

Note: scoring counts a missing base <span class="keyword">as</span> a mismatch, so <span class="keyword">the</span> final
mismatch <span class="command">count</span> <span class="keyword">is</span> <span class="number">2</span> (<span class="number">1</span> '<span class="type">real</span>' mismatch, <span class="number">1</span> 'missing base' mismatch).
If <span class="property">running</span> <span class="keyword">with</span> '<span class="comment">--mismatches 2' (meaning allowing upto 2 mismatches) - this </span>
seqeunce will be classified <span class="keyword">as</span> BC1.
</code></pre><h2 id="Example:_FASTQ_Information">Example: FASTQ Information</h2><h3 id="Genrating_Quality_Information_on_BC54-fq:">Genrating Quality Information on BC54.fq:</h3><p>$ fastx_quality_stats -i BC54.fq -o bc54_stats.txt<br>$ fastq_quality_boxplot_graph.sh -i bc54_stats.txt -o bc54_quality.png -t “My Library”<br>$ fastx_nucleotide_distribution_graph.sh -i bc54_stats.txt -o bc54_nuc.png -t “My Library”</p>
<p></p><p><img src="http://7xk19o.com1.z0.glb.clouddn.com/bc54_quality.png" alt=""><br><img src="http://7xk19o.com1.z0.glb.clouddn.com/bc54_nuc.png" alt=""></p> <p></p>
<h2 id="Example:_FASTQ/A_Manipulation">Example: FASTQ/A Manipulation</h2><h3 id="Common_pre-processing_work-flow:">Common pre-processing work-flow:</h3><p>Covnerting FASTQ to FASTA<br>Clipping the Adapter/Linker<br>Trimming to 27nt (if you’re analyzing miRNAs, for example)<br>Collapsing the sequences<br>Plotting the clipping results</p>
<h3 id="Using_the_FASTX-toolkit_from_the_command_line:">Using the FASTX-toolkit from the command line:</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">$ fastq_to_fasta -v -n -i BC54.fq -o BC54.fa</span><br><span class="line">Input: <span class="number">100000</span> reads.</span><br><span class="line">Output: <span class="number">100000</span> reads.</span><br><span class="line">$ fastx_clipper -v -i BC54.fa <span class="operator">-a</span> CTGTAGGCACCATCAATTCGTA -o BC54.clipped.fa</span><br><span class="line">Clipping Adapter: CTGTAGGCACCATCAATTCGTA</span><br><span class="line">Min. Length: <span class="number">15</span></span><br><span class="line">Input: <span class="number">100000</span> reads.</span><br><span class="line">Output: <span class="number">92533</span> reads.</span><br><span class="line">discarded <span class="number">468</span> too-short reads.</span><br><span class="line">discarded <span class="number">6939</span> adapter-only reads.</span><br><span class="line">discarded <span class="number">60</span> N reads.</span><br><span class="line">$ fastx_trimmer -v <span class="operator">-f</span> <span class="number">1</span> <span class="operator">-l</span> <span class="number">27</span> -i BC54.clipped.fa -o BC54.trimmed.fa</span><br><span class="line">Trimming: base <span class="number">1</span> to <span class="number">27</span></span><br><span class="line">Input: <span class="number">92533</span> reads.</span><br><span class="line">Output: <span class="number">92533</span> reads.</span><br><span class="line">$ fastx_collapser -v -i BC54.trimmed.fa -o BC54.collapsed.fa</span><br><span class="line">Collapsd <span class="number">92533</span> reads into <span class="number">36431</span> unique sequences.</span><br><span class="line">$ fasta_clipping_histogram.pl BC54.collapsed.fa bc54_clipping.png</span><br></pre></td></tr></table></figure>
<h3 id="Mapping_(or_any_other_kind_of_analysis)_of_the_Clipped_+_Collapsed_FASTA_file_will_be:">Mapping (or any other kind of analysis) of the Clipped + Collapsed FASTA file will be:</h3><p>quicker - each unique sequence appears only once in the FASTA file.<br>more accurate - the Adapter/Linker sequence was removed from the 3’ end, and will affect the mapping results.</p>

      
    </div>

    <div>
      
        
<div id="wechat_subscriber" style="display: block; padding: 10px 0; margin: 20px auto; width: 100%; text-align: center">
    <img id="wechat_subscriber_qcode" src="/images/Wechat.jpg" alt="tiramisutes wechat" style="width: 200px; max-width: 100%;"/>
    <div>欢迎关注</div>
</div>


      
    </div>

    <div>
      
        

      
    </div>

    <div>
      
        
  <ul class="post-copyright">
    <li class="post-copyright-author">
      <strong>本文作者:</strong>
      tiramisutes
    </li>
    <li class="post-copyright-link">
      <strong>本文固定链接:</strong>
      <a href="http://tiramisutes.github.io/2015/07/24/FASTX-Toolkit.html" title="FASTX-Toolkit">http://tiramisutes.github.io/2015/07/24/FASTX-Toolkit.html</a>
      <strong>转载请注明出处！</strong>
    </li>
    <li class="post-copyright-license">
      <strong>Copyright Notice: </strong>
      All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/3.0/" rel="external nofollow" target="_blank">CC BY-NC-SA 3.0</a> unless stating additionally.
    </li>
  </ul>


      
    </div>

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/FASTX/" rel="tag"># FASTX</a>
          
        </div>
      

      
        
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2015/07/24/hello-world.html" rel="next" title="Hello World">
                <i class="fa fa-chevron-left"></i> Hello World
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
              <a href="/2015/07/25/Linux-sort.html" rel="prev" title="Linux-sort">
                Linux-sort <i class="fa fa-chevron-right"></i>
              </a>
            
          </div>
        </div>
      

      
      
    </footer>
  </article>



    <div class="post-spread">
      
        <!-- JiaThis Button BEGIN -->
<div class="jiathis_style">
  <a class="jiathis_button_tsina"></a>
  <a class="jiathis_button_tqq"></a>
  <a class="jiathis_button_weixin"></a>
  <a class="jiathis_button_cqq"></a>
  <a class="jiathis_button_douban"></a>
  <a class="jiathis_button_renren"></a>
  <a class="jiathis_button_qzone"></a>
  <a class="jiathis_button_kaixin001"></a>
  <a class="jiathis_button_copy"></a>
  <a href="http://www.jiathis.com/share" class="jiathis jiathis_txt jiathis_separator jtico jtico_jiathis" target="_blank"></a>
  <a class="jiathis_counter_style"></a>
</div>
<script type="text/javascript" >
  var jiathis_config={
    hideMore:false
  }
</script>
<script type="text/javascript" src="http://v3.jiathis.com/code/jia.js" charset="utf-8"></script>
<!-- JiaThis Button END -->

      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
      <div id="hypercomments_widget"></div>
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap" >
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            Overview
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image"
               src="/images/author.jpg"
               alt="tiramisutes" />
          <p class="site-author-name" itemprop="name">tiramisutes</p>
           
              <p class="site-description motion-element" itemprop="description"></p>
           
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives">
                <span class="site-state-item-count">129</span>
                <span class="site-state-item-name">posts</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-categories">
              <a href="/categories/index.html">
                <span class="site-state-item-count">11</span>
                <span class="site-state-item-name">categories</span>
              </a>
            </div>
          

          
            
            
            <div class="site-state-item site-state-tags">
              <a href="/tags/index.html">
                <span class="site-state-item-count">100</span>
                <span class="site-state-item-name">tags</span>
              </a>
            </div>
          

        </nav>

        
          <div class="feed-link motion-element">
            <a href="/atom.xml" rel="alternate">
              <i class="fa fa-rss"></i>
              RSS
            </a>
          </div>
        

        <div class="links-of-author motion-element">
          
            
              <span class="links-of-author-item">
                <a href="https://github.com/tiramisutes" target="_blank" title="GitHub">
                  
                    <i class="fa fa-fw fa-github"></i>
                  
                  GitHub
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="https://twitter.com/xzpgocxx" target="_blank" title="Twitter">
                  
                    <i class="fa fa-fw fa-twitter"></i>
                  
                  Twitter
                </a>
              </span>
            
              <span class="links-of-author-item">
                <a href="http://weibo.com/tiramisutesTes" target="_blank" title="Weibo">
                  
                    <i class="fa fa-fw fa-weibo"></i>
                  
                  Weibo
                </a>
              </span>
            
          
        </div>

        
        

        
        
          <div class="links-of-blogroll motion-element links-of-blogroll-inline">
            <div class="links-of-blogroll-title">
              <i class="fa  fa-fw fa-globe"></i>
              Links
            </div>
            <ul class="links-of-blogroll-list">
              
                <li class="links-of-blogroll-item">
                  <a href="https://tiramisutes.gitbooks.io/awk/content/" title="GitBook《awk学习总结》" target="_blank">GitBook《awk学习总结》</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://cotton.hzau.edu.cn/" title="棉花遗传改良团队" target="_blank">棉花遗传改良团队</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.bio-info-trainee.com/" title="生信菜鸟团-深圳大学城" target="_blank">生信菜鸟团-深圳大学城</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.biotrainee.com/forum.php" title="生信技能树" target="_blank">生信技能树</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.iwhgao.com/" title="rabbit gao's blog" target="_blank">rabbit gao's blog</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://huans.github.io/" title="Huans(代谢+GWAS)" target="_blank">Huans(代谢+GWAS)</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.sequencing.uio.no/" title="NSC-sequenceing" target="_blank">NSC-sequenceing</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://blog.jobbole.com/" title="伯乐在线" target="_blank">伯乐在线</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://linux.cn/" title="Linux开源中文社区" target="_blank">Linux开源中文社区</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://www.r-bloggers.com/" title="R-bloggers" target="_blank">R-bloggers</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="https://cran.r-project.org/other-docs.html" title="R-resource" target="_blank">R-resource</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html" title="ggplot2" target="_blank">ggplot2</a>
                </li>
              
                <li class="links-of-blogroll-item">
                  <a href="http://lilibei.net/" title="pele' blog" target="_blank">pele' blog</a>
                </li>
              
            </ul>
          </div>
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#Command_Line_Arguments"><span class="nav-number">1.</span> <span class="nav-text">Command Line Arguments</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTQ-to-FASTA"><span class="nav-number">1.1.</span> <span class="nav-text">FASTQ-to-FASTA</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTX_Statistics"><span class="nav-number">1.2.</span> <span class="nav-text">FASTX Statistics</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTQ_Quality_Chart"><span class="nav-number">1.3.</span> <span class="nav-text">FASTQ Quality Chart</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA/Q_Nucleotide_Distribution"><span class="nav-number">1.4.</span> <span class="nav-text">FASTA/Q Nucleotide Distribution</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA/Q_Clipper"><span class="nav-number">1.5.</span> <span class="nav-text">FASTA/Q Clipper</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA/Q_Renamer"><span class="nav-number">1.6.</span> <span class="nav-text">FASTA/Q Renamer</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA/Q_Trimmer"><span class="nav-number">1.7.</span> <span class="nav-text">FASTA/Q Trimmer</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA/Q_Collapser"><span class="nav-number">1.8.</span> <span class="nav-text">FASTA/Q Collapser</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTQ/A_Artifacts_Filter"><span class="nav-number">1.9.</span> <span class="nav-text">FASTQ/A Artifacts Filter</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTQ_Quality_Filter"><span class="nav-number">1.10.</span> <span class="nav-text">FASTQ Quality Filter</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTQ/A_Reverse_Complement"><span class="nav-number">1.11.</span> <span class="nav-text">FASTQ/A Reverse Complement</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA_Formatter"><span class="nav-number">1.12.</span> <span class="nav-text">FASTA Formatter</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA_Nucleotides_Changer"><span class="nav-number">1.13.</span> <span class="nav-text">FASTA Nucleotides Changer</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTA_Clipping_Histogram"><span class="nav-number">1.14.</span> <span class="nav-text">FASTA Clipping Histogram</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#FASTX_Barcode_Splitter"><span class="nav-number">1.15.</span> <span class="nav-text">FASTX Barcode Splitter</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Example:_FASTQ_Information"><span class="nav-number">2.</span> <span class="nav-text">Example: FASTQ Information</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Genrating_Quality_Information_on_BC54-fq:"><span class="nav-number">2.1.</span> <span class="nav-text">Genrating Quality Information on BC54.fq:</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Example:_FASTQ/A_Manipulation"><span class="nav-number">3.</span> <span class="nav-text">Example: FASTQ/A Manipulation</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#Common_pre-processing_work-flow:"><span class="nav-number">3.1.</span> <span class="nav-text">Common pre-processing work-flow:</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Using_the_FASTX-toolkit_from_the_command_line:"><span class="nav-number">3.2.</span> <span class="nav-text">Using the FASTX-toolkit from the command line:</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Mapping_(or_any_other_kind_of_analysis)_of_the_Clipped_+_Collapsed_FASTA_file_will_be:"><span class="nav-number">3.3.</span> <span class="nav-text">Mapping (or any other kind of analysis) of the Clipped + Collapsed FASTA file will be:</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright" >
  
  &copy;  2014 - 
  <span itemprop="copyrightYear">2018</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder"> <a href="/about" target="_blank">tiramisutes</a></span>
</div>

<div class="powered-by">
  Powered by <a class="theme-link" href="https://hexo.io" target="_blank">Hexo</a>
</div>

<div class="theme-info">
  Theme -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next" target="_blank">
    NexT.Muse 
  </a>
<script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js">
                </script>
    <span id="busuanzi_container_site_pv" style='display:none'>
                      &nbsp;&nbsp;&nbsp;|&nbsp;&nbsp;&nbsp;<a style="line-height: 1.5;" href="http://c4.gostats.cn/last_guests.xml?id=397056" target="_blank">The total visits <span id="busuanzi_value_site_pv"></span> times</a>
                </span>
    <!-- GoStats JavaScript Based Code -->
                <script type="text/javascript" src="http://gostats.cn/js/counter.js"></script>
                <script type="text/javascript">_gos='c4.gostats.cn';_goa=397056;
                _got=5;_goi=1;_GoStatsRun();</script>
                <noscript><a target="_blank" 
                href="http://gostats.cn"><img src="http://c4.gostats.cn/bin/count/a_397056/t_5/i_1/counter.png" 
                style="border-width:0" /></a></noscript>
                <!-- End GoStats JavaScript Based Code -->  
</div>
<script>
 host3 = window.location.href;
  chinadomai3 = "http://tiramisutes.github.io/";
  if (host3.indexOf(chinadomain3)!=-1)  {

  (function(){
    var bp = document.createElement('script');
    bp.src = '//push.zhanzhang.baidu.com/push.js';
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>
        

        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.1"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.1"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.1"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.1"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.1"></script>



  


  




	

		<script type="text/javascript">
		_hcwp = window._hcwp || [];

		_hcwp.push({widget:"Bloggerstream", widget_id: 91479, selector:".hc-comment-count", label: "{\%COUNT%\}" });

		
		_hcwp.push({widget:"Stream", widget_id: 91479, xid: "2015/07/24/FASTX-Toolkit.html"});
		

		(function() {
		if("HC_LOAD_INIT" in window)return;
		HC_LOAD_INIT = true;
		var lang = (navigator.language || navigator.systemLanguage || navigator.userLanguage || "en").substr(0, 2).toLowerCase();
		var hcc = document.createElement("script"); hcc.type = "text/javascript"; hcc.async = true;
		hcc.src = ("https:" == document.location.protocol ? "https" : "http")+"://w.hypercomments.com/widget/hc/91479/"+lang+"/widget.js";
		var s = document.getElementsByTagName("script")[0];
		s.parentNode.insertBefore(hcc, s.nextSibling);
		})();
		</script>

	












  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (search_path.endsWith("json")) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  
  <script src="https://cdn1.lncld.net/static/js/av-core-mini-0.6.1.js"></script>
  <script>AV.initialize("c7x0fPmO8Ci9avDBRuGWHqpG-gzGzoHsz", "iVJ6ST0e1O5xaiIhVwOyhbf3");</script>
  <script>
    function showTime(Counter) {
      var query = new AV.Query(Counter);
      var entries = [];
      var $visitors = $(".leancloud_visitors");

      $visitors.each(function () {
        entries.push( $(this).attr("id").trim() );
      });

      query.containedIn('url', entries);
      query.find()
        .done(function (results) {
          var COUNT_CONTAINER_REF = '.leancloud-visitors-count';

          if (results.length === 0) {
            $visitors.find(COUNT_CONTAINER_REF).text(0);
            return;
          }

          for (var i = 0; i < results.length; i++) {
            var item = results[i];
            var url = item.get('url');
            var time = item.get('time');
            var element = document.getElementById(url);

            $(element).find(COUNT_CONTAINER_REF).text(time);
          }
          for(var i = 0; i < entries.length; i++) {
            var url = entries[i];
            var element = document.getElementById(url);
            var countSpan = $(element).find(COUNT_CONTAINER_REF);
            if( countSpan.text() == '') {
              countSpan.text(0);
            }
          }
        })
        .fail(function (object, error) {
          console.log("Error: " + error.code + " " + error.message);
        });
    }

    function addCount(Counter) {
      var $visitors = $(".leancloud_visitors");
      var url = $visitors.attr('id').trim();
      var title = $visitors.attr('data-flag-title').trim();
      var query = new AV.Query(Counter);

      query.equalTo("url", url);
      query.find({
        success: function(results) {
          if (results.length > 0) {
            var counter = results[0];
            counter.fetchWhenSave(true);
            counter.increment("time");
            counter.save(null, {
              success: function(counter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(counter.get('time'));
              },
              error: function(counter, error) {
                console.log('Failed to save Visitor num, with error message: ' + error.message);
              }
            });
          } else {
            var newcounter = new Counter();
            /* Set ACL */
            var acl = new AV.ACL();
            acl.setPublicReadAccess(true);
            acl.setPublicWriteAccess(true);
            newcounter.setACL(acl);
            /* End Set ACL */
            newcounter.set("title", title);
            newcounter.set("url", url);
            newcounter.set("time", 1);
            newcounter.save(null, {
              success: function(newcounter) {
                var $element = $(document.getElementById(url));
                $element.find('.leancloud-visitors-count').text(newcounter.get('time'));
              },
              error: function(newcounter, error) {
                console.log('Failed to create');
              }
            });
          }
        },
        error: function(error) {
          console.log('Error:' + error.code + " " + error.message);
        }
      });
    }

    $(function() {
      var Counter = AV.Object.extend("Counter");
      if ($('.leancloud_visitors').length == 1) {
        addCount(Counter);
      } else if ($('.post-title-link').length > 1) {
        showTime(Counter);
      }
    });
  </script>



  
<script>
(function(){
    var bp = document.createElement('script');
    var curProtocol = window.location.protocol.split(':')[0];
    if (curProtocol === 'https') {
        bp.src = 'https://zz.bdstatic.com/linksubmit/push.js';        
    }
    else {
        bp.src = 'http://push.zhanzhang.baidu.com/push.js';
    }
    var s = document.getElementsByTagName("script")[0];
    s.parentNode.insertBefore(bp, s);
})();
</script>


  
  


  
  <script type="text/javascript" src="/js/src/js.cookie.js?v=5.1.1"></script>
  <script type="text/javascript" src="/js/src/scroll-cookie.js?v=5.1.1"></script>


  

</body>
</html>
